--- 
  title: "Salmo smoltification study"
output: plot_heatmap.html
author: Joe Colgan (joscolgan)
---
  
  ## Introduction:  
  This script generates a heatmap for significant differentially expressed genes identified across three experimental groups of brown trout that differ in their phenotype.
The script takes raw gene-level counts generated by [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html), scales counts, subsets rows by most significant genes, clusters based by expression similarity across genes and generates a heatmap.  
The output is a heatmap as .png, .jpeg and/or .pdf.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "reshape2",
               "DESeq2",
               "lintr")
for (lib in libraries) {
  if (require(package = lib, character.only = TRUE)) {
    print("Successful")
  } else {
    print("Installing")
    source("https://bioconductor.org/biocLite.R")
    library(lib, character.only = TRUE)
  }
}
## Create output directory:
dir.create(path = "results")
```

Load data:
  
  ```{r, message = FALSE}
load(file = "DESeq2_output_liver.RData")
load(file = "eulerr_output_liver.RData")
```

Examine count data:
  
  ```{r, message = FALSE}
## Normalise counts:
normalised_counts <- t(scale(t(txi$counts),
                             center = TRUE,
                             scale = TRUE))
## Convert to dataframe:
normalised_counts_df <- as.data.frame(normalised_counts)
## Read in sample names and assign to columns of dataframes:
sample_info <- read.table(file = "data/RNAseq_Sample_File_Reduced.txt",
                          header = TRUE)
## Extract liver samples:
sample_info <- subset(x = sample_info,
                      tissue == "Liver")
## Update name for plotting:
## Reorder based on phenotype:
salt_non_smolt_samples  <- c(1, 2, 5, 6, 9, 10, 11, 14)
salt_smolt_samples      <- c(12, 13, 3, 4, 7, 8)
fresh_non_smolt_samples <- c(16,17,19,20,21,15,18,22)


sample_info_ordered <- sample_info[c(1, 2, 5, 6, 9, 10, 11, 14, 12, 13, 3, 4, 7, 8, 16,17,19,20,21,15,18,22),]

sample_info_ordered

## Reorder the columns by phenotype:
normalised_counts_reordered <- normalised_counts_df[, c(salt_non_smolt_samples,
                                                        salt_smolt_samples,
                                                        fresh_non_smolt_samples)]
## Update sample names:
resident_sw <- "resident-SW"
smolt_sw    <- "smolt-SW"
resident_fw <- "resident-FW"
length(grep(pattern = "salt_NonSmolt",
            colnames(normalised_counts_reordered)))
length(grep(pattern = "salt_Smolt",
            colnames(normalised_counts_reordered)))
length(grep(pattern = "fresh_NonSmolt",
            colnames(normalised_counts_reordered)))
## Generate new sample names for plotting:
sample_names <- c(paste(resident_sw, "_", "male", "_", 1:5,
                        sep = ""),
                  paste(resident_sw, "_", "female", "_", 1:3,
                        sep = ""),
                  paste(smolt_sw, "_", "male", "_", 1:2,
                        sep = ""),
                  paste(smolt_sw, "_", "female", "_", 1:4,
                        sep = ""),  
                  paste(resident_fw, "_", "male", "_", 1:5,
                        sep = ""),
                  paste(resident_fw, "_", "female", "_", 1:3,
                        sep = ""))
                        

## Rename columns:
colnames(normalised_counts_reordered) <- sample_names
```

Subset genes identified as phenotype but NOT environment genes:
  
  ```{r, message = FALSE}

length(phen_AND_env_genes)

normalised_counts_unique_degs <- subset(normalised_counts_reordered,
                                        row.names(normalised_counts_reordered) %in% phen_AND_env_genes)
```

Extract normalised counts for genes of interest:
  
  ```{r, message = FALSE}
## Cluster based on euclidean distance:
dd <- hclust(dist(as.matrix(normalised_counts_unique_degs[, 1:22])))
## Reorder:
normalised_counts_ordered <- as.data.frame(normalised_counts_unique_degs[rev(dd$order), ])
normalised_counts_ordered$locus <- row.names(normalised_counts_ordered)
## Transform dataframe:
normalised_counts_ordered_melt <- melt(normalised_counts_ordered)
## Update column names:
colnames(normalised_counts_ordered_melt) <- c("gene_name",
                                              "sample",
                                              "counts")
## Update levels for plotting:
normalised_counts_ordered_melt$gene_name <-
  factor(normalised_counts_ordered_melt$gene_name,
         levels = rev(unique(c(as.character(unlist(normalised_counts_ordered_melt$gene_name))))))
```

Plot heatmap:
  
  ```{r, message = FALSE}
## Assign the name of dataframe to plot:
input <- normalised_counts_ordered_melt
## Remove underscores in sample names:
input$sample <- gsub(pattern = "_",
                     replacement = " ",
                     input$sample)
## Update levels for plotting:
input$sample <- factor(input$sample,
                       levels = unique(input$sample))
## Generate heatmap:
heatmap_plot <- ggplot(input,
                       aes(sample,
                           gene_name)) +
  geom_tile(aes(fill = counts)) +
  geom_vline(xintercept = c(0.5,
                            8.5,
                            14.5,
                            22.5)) +
  scale_fill_gradient2(low = "black",
                       mid = "white",
                       high = "red") +
  ylab("") +
  xlab("") +
  theme(legend.title = element_text(size = 10,
                                    face = "bold"),
        legend.text = element_text(size = 10,
                                   face = "bold"),
        axis.text.x = element_text(size = 8,
                                   face = "bold",
                                   colour = "black",
                                   angle = 90,
                                   hjust = 1),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        #axis.text.y = element_text(size = 6,
        #                           face = "plain",
        #                           colour = "black"),
        axis.title = element_text(size = 10,
                                  face = "bold"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        legend.position = "bottom") +
  scale_y_discrete(position = "right") +
  labs(fill = "Normalised gene expression")
## Print to console:
heatmap_plot

## Save image:
output  <- "results"
ggsave(file = paste(output,
                    "/",
                    "heatmap_liver_phen_and_env_genes.png",
                    sep = ""),
       height = 6,
       width = 10)
```

Run lintr:
  
  ```{r, message = FALSE}
lintr::lint(file = "plot_heatmap_phen_and_env_genes.Rmd")
```

###########################################################
# Next bits are about generating a nice plot with means (as points) and SEs:

phenotype<- c(rep("resident-SW", 8), rep("smolt-SW", 6), rep("resident-FW", 8))
input$phenotype<- rep(phenotype, each = length(phen_AND_env_genes))

sex<- c(rep("male", 5), rep("female", 3), rep("male", 2), rep("female", 4),rep("male", 5), rep("female", 3))
input$sex <- rep(sex, each = length(phen_AND_env_genes))



# Plot:

par(mfrow=c(1,3), mar=c(4,0,0,0), oma=c(1,2,2,1))

# Plot mean expression of each gene for residents-SW, splitting by sex:
resSW_male<- subset(input, phenotype=="resident-SW" & sex=="male")
y<- with(resSW_male, tapply(counts, gene_name, mean))
x <- 1:length(y)
plot(x~y, pch=21, col="black", bg="green",  xlab="", ylab="Gene index", xlim=c(-2,4))

resSW_female<- subset(input, phenotype=="resident-SW" & sex=="female")
y<- with(resSW_female, tapply(counts, gene_name, mean))
x <- 1:length(y)
points(x~y, pch=1)
mtext("resident-SW")


# Plot mean expression of each gene for smolts-SW, splitting by sex:
smoltSW_male<- subset(input, phenotype=="smolt-SW" & sex=="male")
y<- with(smoltSW_male, tapply(counts, gene_name, mean))
x <- 1:length(y)
plot(x~y, pch=21, col="black", bg="green", xlab="", ylab="Gene index", xlim=c(-2,4))

smoltSW_female<- subset(input, phenotype=="smolt-SW" & sex=="female")
y<- with(smoltSW_female, tapply(counts, gene_name, mean))
x <- 1:length(y)
mtext("smolts-SW")
points(x~y, pch=1)

# Plot mean expression of each gene for residents-FW, splitting by sex:
resFW_male<- subset(input, phenotype=="resident-FW" & sex=="male")
y<- with(resFW_male, tapply(counts, gene_name, mean))
x <- 1:length(y)
plot(x~y, pch=21, col="black", bg="green",  xlab="", ylab="Gene index", xlim=c(-2,4))

resFW_female<- subset(input, phenotype=="resident-FW" & sex=="female")
y<- with(resFW_female, tapply(counts, gene_name, mean))
x <- 1:length(y)
points(x~y, pch=1)
mtext("resident-FW")

################################################################################
## Based on these plots, we can see there are 3 types of genes within these "phen_AND_env_genes":

# 1 = The top block of genes are osmotic environment response genes
# 2 = The middle block of genes are stress genes
# 3 = The bottom block of genes are ALH genes

# So now let's extract these as 3 separate lists. First, we need to find the break-points between the blocks. The break between the top and middle blocks is most clearly seen in the resident-SW group for males:

y<- with(resSW_male, tapply(counts, gene_name, mean))

# Confusingly, y here is ordered in the opposite way to the genes in the heatmap. So the first gene in y is gene 154, and the last gene in y is gene 1.  So we need to look at it backwards!:

rev(y)

# the first element in rev(y) now corresponds to gene 1, and the last element to gene 154.  Scrolling down through rev(y), you can see a clear change from negative to positive values after ENSSTUG00000044969.1 gene. We can find the position of the gene in the list like this:
x<- row.names(rev(y))
which(x=="ENSSTUG00000044969.1")

# So genes 1 to 57 in rev(y) correspond to the top block of genes, and we can extract their names as a list like this:
top_block_genes_phen_AND_env <- x[1:57]

# The break between the middle and top blocks is most clearly seen in the smolt-SW group for females:

y<- with(smoltSW_female, tapply(counts, gene_name, mean))
# Again we need to reverse the order:
rev(y)

# Scrolling down through rev(y), you can see a clear change from negative to positive values after ENSSTUG00000004291.1 gene.
x<- row.names(rev(y))
which(x=="ENSSTUG00000004291.1")

# So genes 58 to 96 in rev(y) correspond to the middle block of genes:
middle_block_genes_phen_AND_env <- x[58:96]

# Leaving the rest as the bottom block:
bottom_block_genes_phen_AND_env <- x[97:154]

# Let's look at mean gene expression for each block to make sure the above worked:

resSW <- subset(input, phenotype=="resident-SW")
mean(resSW$counts[resSW$gene_name %in% top_block_genes_phen_AND_env])
mean(resSW$counts[resSW$gene_name %in% middle_block_genes_phen_AND_env])
mean(resSW$counts[resSW$gene_name %in% bottom_block_genes_phen_AND_env])

smoltsSW <- subset(input, phenotype=="smolt-SW")
mean(smoltsSW$counts[smoltsSW$gene_name %in% top_block_genes_phen_AND_env])
mean(smoltsSW$counts[smoltsSW$gene_name %in% middle_block_genes_phen_AND_env])
mean(smoltsSW$counts[smoltsSW$gene_name %in% bottom_block_genes_phen_AND_env])

resFW <- subset(input, phenotype=="resident-FW")
mean(resFW$counts[resFW$gene_name %in% top_block_genes_phen_AND_env])
mean(resFW$counts[resFW$gene_name %in% middle_block_genes_phen_AND_env])
mean(resFW$counts[resFW$gene_name %in% bottom_block_genes_phen_AND_env])


saveRDS(top_block_genes_phen_AND_env, file="top_block_genes_phen_AND_env.rds")
saveRDS(middle_block_genes_phen_AND_env, file="middle_block_genes_phen_AND_env.rds")
saveRDS(bottom_block_genes_phen_AND_env, file="bottom_block_genes_phen_AND_env.rds")


