
## Here include some basic information
## Name of author:Robert Wynne
## Date:24/01/2020
## Program name: rscript_rna_seq-1.R
## Purpose of the script:RNAseq analysis
## This script does XX, YYY and ZZZ.
##
## It takes salmon output from ICHEC as input, performs YYY and outputs ZZZ.
## it takes transcription quantification files
## generated by the pseudoaligner salmon, generates
## gene-level counts using Tximport and performs
## differential expression analyses using DESeq2.

##It outputs... - We can work on this when I get back.
## E.g. tables of differentially expressed genes, PCA
## plots.

#RNA seq analysis
```{r, message = FALSE}
# Install libraries:
#install required packages,
#must be done in this way to assure all of the packages work correctly

#if (!requireNamespace("BiocManager", quietly = TRUE))
# install.packages("BiocManager")
##BiocManager::install(version = "3.10")
#BiocManager::install("DESeq2")
#BiocManager::install("apeglm")
#BiocManager::install("tximportData")
#BiocManager::install("DESeqDataSetFromTximport")
#install.packages("ggplot2")
```

Load libraries:

```{r, message = FALSE}
## Load libraries:
library(apeglm)
library(ggplot2)
library(DESeq2)
library(RCurl)
library(stringi)
library(tximport)

# Check version using sessionInfo
# E.g. sessionInfo("DESeq2")
```

Read in input files. The input files for this script are generated by the pseudoaligner, salmon, which quantifies expression of individual transcript. For differential expression analysis, we want to generate gene-level counts for investigating differences in gene expression between one or more treatment groups.
 
```{r, message = FALSE}
## Check working directory and read in input files:
##  If your script is in the same directory as your input files
## you don't need to set the working directory - remove if redundant.
#set working directory
setwd("~/")
#make sure we are working in the correct directory
getwd()

# read in salmon output files
## JC - This step reads in a list of sample names rather than the actual files
## JC - Don't use capital letters in variable names - 'RNAsamples' should
## be 'rna_samples'
## JC - Actually state what is in 'RNAseq_Sample_File_Reduced.txt' - how many
## columns and what they relate to i.e. First column contains sample names;
## Second column contains the sex of the samples; etc.
rna_sample_info <- read.table(file = "./RNASEQ/RNAseq_Sample_File_Reduced.txt", header = TRUE)
#check the data
## JC - See comment above:
head(rna_sample_info)

## RNASeq analysis for liver samples:
# give R the file path needed, we do this by taking any file from the dirctory with
#'quant.sf attached to the end of the name
## JC - Create a vector containing a list of paths for each sample
## listed in the sample information file read in and stored as 'rna_sample_info'.
## The function 'file.path' will look in the directory 'RNASEQ' and then within indivdiual folders contained within that directory. Each of the individual folders in that directory should contain folders for each sample listed in the sample information file. Within each of these folders, files all end in the file extension 'quant.sf', which is the
## default naming scheme by Salmon.
files <- file.path("./RNASEQ",
                   rna_sample_info$sample,
                   "quant.sf")

# At present, all samples are being loaded in as input, which contains a mix of liver and brain samples.
#create a dataframe which labels all of the Liver files as true
#this will help is separate the organs as we will be looking at the
#brain and liver separetly
files_df <- as.data.frame(cbind(files,
                                grepl(pattern = "_L",
                                      x = files)))

## creating a subset of just liver samples
## JC: This step subsets the dataframe 'files_df' by checking if values in column 'V2'
## are annotated as 'TRUE'. If so, it subsets these entire lines into a new dataframe
## called 'files_liver_df'.
files_liver_df <- subset(files_df,
                         V2 == TRUE)

## creating a subset with all of the discriptive information for all of the liver samples
##from the 'RNAseqsamples' document
## JC - Again, 'subset' is the function, it takes dataframes as input and creates dataframes
## as output. For this command, the dataframe 'RNAsamples' is subset by the column which contains
## 'tissue' information and only rows related to samples annotated as 'Liver' are subsetted into
## a new dataframe 'RNAsamples_liver'.
rna_sample_info_liver <- subset(rna_sample_info,
                                tissue == "Liver")


## CHECK IF REDUNDANT - IF SO REMOVE
## create a factor from the sample file.
#file <- files_liver_df$files

## create a list of sample names.
#names(file) <- RNAsamples_liver$sample
#names(file)

## specify the path to the files using the appropriate columns of samples
## also need to link transcripts to genes for this dataset:
tx2gene <- read.table("./RNASEQ/combined_transcript_to_gene.txt")

## Ensure that names are of class character:
files <- as.character(files_liver_df$files)

## Generate gene level counts:
txi <- tximport(files,
                type = "salmon",
                tx2gene = tx2gene)

#check class to make sure data is a list
class(txi)

## Redundant?
## construct a DESeqDataSet from the txi object and sample information in samples.
#dds_txi <- DESeqDataSetFromTximport(txi,
#                                   colData = rna_sample_info_liver,
#                                   design = ~ environment + maturity)

## Possibly redundant:
# dds <- DESeq(dds_txi)

### group factors for different comparasons
rna_sample_info_liver$group <- paste(rna_sample_info_liver$environment,
                                     "_",
                                     rna_sample_info_liver$phenotype,
                                     sep = "")

### input values and create intermediate calculations and results of an analysis of differential expression looking at sex and
### 'group'(a constructed column combining phenotype and enviornment).
dds_txi <- DESeqDataSetFromTximport(txi,
                                   colData = RNAsamples_liver,
                                   design = ~sex + group)

### using the results from above to run the DESeq analysis
dds <- DESeq(ddsTxi,
             test = "LRT",
             reduced = ~sex)
```

Compare numbers of differentially expressed genes between treatments of interest:
1) Smolt saltwater vs Non-smolt saltwater
2) Smolt saltwater vs Non-smolt freshwater
3) Non-smolt freshwater vs Non-smolt saltwater

```{r, message = FALSE}
## JC: Here you are constrasting two specific groups contained within the DESeq object (dds).
## The output of this function is a dataframe named 'dds_salt'.

## Extract results for differentially expressed genes:
# 1) Smolt saltwater vs Non-smolt saltwater
dds_salt_s_salt_ns <- results(dds,
                              contrast = c("group",
                                      "salt_NonSmolt",
                                      "salt_Smolt"))

# 2) Smolt saltwater vs Non-smolt freshwater
dds_salt_s_fresh_ns <- results(dds,
                               contrast = c("group",
                                      "salt_NonSmolt",
                                      "salt_Smolt"))

# 3) Non-smolt freshwater vs Non-smolt saltwater:
dds_fresh_ns_salt_ns <- results(dds,
                                contrast = c("group",
                                             "salt_NonSmolt",
                                             "salt_Smolt"))
```

We also want to perform Wald test between comparisons and take genes identified by both approaches as differentially expressed between groups as indentified by the LRT.

```{r, message = FALSE}
### Run a Wald test:
dds_wald <- DESeq(dds_txi,
             test = "Wald",
             prior = TRUE)
```

Extract group comparisons similar to above for the LRTs:
Compare numbers of differentially expressed genes between treatments of interest:
1) Smolt saltwater vs Non-smolt saltwater
2) Smolt saltwater vs Non-smolt freshwater
3) Non-smolt freshwater vs Non-smolt saltwater

```{r, message = FALSE}
## compare smolts and non smolts
dds_salt_smolt_vs_salt_non_smolt <- results(dds_wald,
                                            contrast = c("group",
                                                         "salt_NonSmolt",
                                                         "salt_Smolt"))


dds_salt_smolt_vs_fresh_non_smolt <- results(dds_wald,
                                            contrast = c("group",
                                                         "fresh_NonSmolt",
                                                         "salt_Smolt"))


dds_salt_non_smolt_vs_fresh_non_smolt <- results(dds_wald,
                                             contrast = c("group",
                                                          "fresh_NonSmolt",
                                                          "salt_NonSmolt"))
```

Count the number of differentally expressed genes between the groups

```{r, message = FALSE}
## create dataframe of gene list, subsetting the p.value
##count rows
nrow(dds_salt_smolt_vs_salt_non_smolt)
nrow(dds_salt_non_smolt_vs_fresh_non_smolt)
nrow(dds_non_smolt_salt_vs_non_smolt_fresh)

## The 'results' function of DESeq2 outputs a table with data format similar to a dataframe but
## not exactly the same. Therefore, convert these files into dataframes:
smolt_vs_non_smolts_salt_df <- as.data.frame(dds_salt_smolt_vs_salt_non_smolt)
smolt_vs_non_smolts_fresh_df <- as.data.frame(dds_salt_smolt_vs_fresh_non_smolt)
non_smolt_salt_vs_non_smolts_fresh_df <- as.data.frame(dds_salt_smolt_vs_fresh_non_smolt)
```

Subset significant genes for each comparison:

```{r, message = FALSE}
smolt_vs_non_smolts_salt_sig  <- subset(smolt_vs_non_smolts_salt_df,
                                        padj < 0.05)
smolt_vs_non_smolts_fresh_sig <- subset(smolt_vs_non_smolts_fresh_df,
                                        padj < 0.05)
non_smolt_salt_vs_non_smolts_fresh_sig <- subset(smolt_vs_non_smolts_fresh_df,
                                                 padj < 0.05)
```

Subset gene name and p value for all genes analysed:

```{r, message = FALSE}
## For each comparison, extract:
## 1) gene name (taken from the row names of each dataframe) # Required for Euler plots
## 2) unadjusted p value # Required for GO term enrichment
## Comparison 1:
smolt_vs_non_smolts_salt_gene_pvalue <- as.data.frame(cbind(row.names(smolt_vs_non_smolts_salt_df),
                                                            smolt_vs_non_smolts_salt_df[, 5:6]))
## Update column name:
colnames(smolt_vs_non_smolts_salt_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")

## Comparison 2:
smolt_vs_non_smolts_fresh_gene_pvalue <- as.data.frame(cbind(row.names(smolt_vs_non_smolts_fresh_df),
                                                            smolt_vs_non_smolts_fresh_df[, 5:6]))
## Update column name:
colnames(smolt_vs_non_smolts_fresh_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")

## Comparison 3:
non_smolt_salt_vs_non_smolts_fresh_gene_pvalue <- as.data.frame(cbind(row.names(non_smolt_salt_vs_non_smolts_fresh_df),
                                                            non_smolt_salt_vs_non_smolts_fresh_df[, 5:6]))
## Update column name:
colnames(non_smolt_salt_vs_non_smolts_fresh_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")
```

Save everyting:

```{r, message = FALSE}
save.image(file = "./DESeq2_output_liver.RData")
```

Generate PCA plots to explore expression profiles amongst individuals and treatments:

```{r, message = FALSE}
## JC: Transform data for plotting:
vsd <- vst(dds,
           blind = FALSE)

## Generate plot:
plotPCA(vsd,
        intgroup=c("group")) +
        ggtitle("Varience between the liver transcriptome profiles of called phenotypes in different Environments") +
        geom_text(label = rna_sample_info_liver$sample,
            nudge_x = 3)


plotPCA(vsd, intgroup=c("phenotype", "sex"))
plotPCA(vsd, intgroup=c("sex"))
plotPCA(vsd, intgroup=c("phenotype"))

plotPCA(vsd, intgroup=c("environment"))
plotPCA(vsd, intgroup=c("maturity"))
```

Save everything:

```{r, message = FALSE}
save.image(file = "./DESeq2_output_liver.RData")
```
