--- 
title: "Salmo smolt study"
output: sex_bias_brain_script.html
authors: Robert Wynne, Tom Reed, Joe Colgan (joscolgan)
---
  
1. Load libraries:
  
```{r, message = FALSE}
## Load libraries:
library(apeglm)
library(ggplot2)
library(DESeq2)
library(RCurl)
library(stringi)
library(tximport)
library(biomaRt)
# Check version using sessionInfo
# E.g. sessionInfo("DESeq2")
```

2. Read in input files. The input files for this script are generated by the pseudoaligner, salmon, which quantifies expression of individual transcript. For differential expression analysis, we want to generate gene-level counts for investigating differences in gene expression between one or more treatment groups.

```{r, message = FALSE}
## Check working directory and read in input files:
##  If your script is in the same directory as your input files
## you don't need to set the working directory - remove if redundant.
#set working directory
setwd("~/")
#make sure we are working in the correct directory
getwd()
## Load R object:
load(file = "../results/figures_for_ms/DESeq2_output_brain.RData")
```

3. Create model for investigating interaction effect between group and sex:
  
```{r, message = FALSE}
### input values and create intermediate calculations and results of an analysis of differential expression looking at sex and
### 'group'(a constructed column combining phenotype and enviornment).
dds_sex <- DESeqDataSetFromTximport(txi,
                                    colData = rna_sample_info_brain,
                                    design = ~sex)
### using the results from above to run the DESeq analysis
dds_null <- DESeq(dds_sex,
                  test = "LRT",
                  reduced = ~ 1)
```

4. Investigate main effect of sex:
  
```{r, message = FALSE}
options(scipen = 999)

## Check names of comparisons:
resultsNames(dds_null)

## Extract significant results:
sex_effect <- results(dds_null,
                      name = "sex_Male_vs_Female")

## Convert to dataframe:
sex_effect_df <- as.data.frame(sex_effect)

## Add gene description:
sex_effect_df$description <- salmo_description[match(row.names(sex_effect_df),
                                                     salmo_description$ensembl_gene_id), ]

## Write to file:
write.table(x = sex_effect_df,
            file = "results/sex_effect_results_brain_annotated.txt",
            col.names = TRUE,
            row.names = TRUE,
            quote = FALSE,
            sep = "\t")

## Calculate absolute values:
sex_effect$absL2FC <- abs(sex_effect$log2FoldChange)

## Extract significant genes:
sex_effect_sig <- subset(x = sex_effect,
                         padj < 0.05 & absL2FC > 1)

## Get gene list:
sex_sig_genes <- row.names(sex_effect_sig)
length(sex_sig_genes)
```

5a. Check overlap with genes that differ between alternative life-stage phenotypes but not due to environment:

```{r, message = FALSE}
## Check overlap:
putative_alh_genes <- intersect(sex_sig_genes,
                                phen_not_env_genes)

## Count the number of genes that overlap:
length(putative_alh_genes)

## Calculate the percentage of ALH genes with sex-biased expression:
length(putative_alh_genes) / length(phen_not_env_genes)
```

5b. Check overlap with genes that differ between alternative life-stage phenotypes and also environment:

```{r, message = FALSE}
## Check overlap:
putative_stress_genes <- intersect(sex_sig_genes,
                                phen_and_env_genes)

## Count the number of genes that overlap:
length(putative_stress_genes)

## Calculate the percentage of ALH genes with sex-biased expression:
length(putative_stress_genes) / length(phen_and_env_genes)
```

5c. Check overlap with genes that don't differ between alternative life-stage phenotypes but only between environments:

```{r, message = FALSE}
## Check overlap:
putative_env_genes <- intersect(sex_sig_genes,
                                env_not_phen_genes)

## Count the number of genes that overlap:
length(putative_env_genes)

## Calculate the percentage of ALH genes with sex-biased expression:
length(putative_env_genes) / length(env_not_phen_genes)
```

6. Implement chi-squared test:
a) For sex-biased genes also identified as ALH genes:

```{r, message = FALSE}
## Calculate the number of sex-biased genes with no overlap with ALH genes:
non_alh_sex_effect_genes <- setdiff(sex_sig_genes,
                                   putative_alh_genes)
length(non_alh_sex_effect_genes)

## Extract the number of total genes expressed in the brown trout transcriptome:
all_genes <- row.names(env_effect_contr_phen)
length(all_genes)
length(non_alh_sex_effect_genes) / (length(all_genes) - length(putative_alh_genes))

# Let's now make a contingency table of ALH/non-ALH genes versus sex-biased/unbiased genes:
x <- as.table(rbind(c(length(putative_alh_genes),
                      length(non_alh_sex_effect_genes)),
                    c((length(phen_not_env_genes) - length(putative_alh_genes)),
                      (length(all_genes) - length(phen_not_env_genes) - length(non_alh_sex_effect_genes)))))

## Rename row and column names:
row.names(x) <- c("Sex-biased",
                 "Unbiased")
colnames(x)  <- c("ALH genes",
                  "Non-ALH genes")

## Print to console:
x

# The total needs to sum to length(all_genes):
sum(x)
length(all_genes)

# And now a chi-square test:
chiseq_res <- chisq.test(x)

## Print to console the observed and expected values:
chiseq_res$observed
chiseq_res$expected
```

b) For sex-biased genes also identified as stress (phenotype + environment) genes:

```{r, message = FALSE}
## Calculate the number of sex-biased genes with no overlap with stress genes:
non_stress_sex_effect_genes <- setdiff(sex_sig_genes,
                                   putative_stress_genes)
length(non_stress_sex_effect_genes)

## Extract the number of total genes expressed in the brown trout transcriptome:
all_genes <- row.names(env_effect_contr_phen)
length(all_genes)
length(non_stress_sex_effect_genes) / (length(all_genes) - length(putative_stress_genes))

# Let's now make a contingency table of stress/non-stress genes versus sex-biased/unbiased genes:
x_stress <- as.table(rbind(c(length(putative_stress_genes),
                      length(non_stress_sex_effect_genes)),
                    c((length(phen_and_env_genes) - length(putative_stress_genes)),
                      (length(all_genes) - length(phen_and_env_genes) - length(non_stress_sex_effect_genes)))))

## Rename row and column names:
row.names(x_stress) <- c("Sex-biased",
                         "Unbiased")
colnames(x_stress)  <- c("Stress genes",
                         "Non-stress genes")

## Print to console:
x_stress

# The total needs to sum to length(all_genes):
sum(x_stress)
length(all_genes)

# And now a chi-square test:
chiseq_res_x_stress <- chisq.test(x_stress)

## Print to console the observed and expected values:
chiseq_res_x_stress$observed
chiseq_res_x_stress$expected
```

c) For sex-biased genes also identified as environment (not phenotype but only environment) genes:

```{r, message = FALSE}
## Calculate the number of sex-biased genes with no overlap with stress genes:
non_env_sex_effect_genes <- setdiff(sex_sig_genes,
                                   putative_env_genes)
length(non_env_sex_effect_genes)

## Extract the number of total genes expressed in the brown trout transcriptome:
all_genes <- row.names(env_effect_contr_phen)
length(all_genes)
length(non_env_sex_effect_genes) / (length(all_genes) - length(putative_env_genes))

# Let's now make a contingency table of stress/non-stress genes versus sex-biased/unbiased genes:
x_env <- as.table(rbind(c(length(putative_env_genes),
                      length(non_env_sex_effect_genes)),
                    c((length(env_not_phen_genes) - length(putative_env_genes)),
                      (length(all_genes) - length(env_not_phen_genes) - length(non_env_sex_effect_genes)))))

## Rename row and column names:
row.names(x_env) <- c("Sex-biased",
                      "Unbiased")
colnames(x_env)  <- c("Environent-response genes",
                      "Non-environent-response genes")

## Print to console:
x_env

# The total needs to sum to length(all_genes):
sum(x_env)
length(all_genes)

# And now a chi-square test:
chiseq_res_x_env <- chisq.test(x_env)

## Print to console the observed and expected values:
chiseq_res_x_env$observed
chiseq_res_x_env$expected
```

7. Save everything:

```{r, message = FALSE}
save.image(file = "DESeq2_output_sex_effects_liver.RData")
```

8. Run lintr:

```{r, message = FALSE}
lintr::lint(file = "sex_bias_brain_script.Rmd")
```
