--- 
title: "Salmo smoltification study"
output: plot_heatmap.html
author: Joe Colgan (joscolgan)
---
  
## Introduction:  
This script generates a heatmap for significant differentially expressed genes identified across three experimental groups of brown trout that differ in their phenotype.
The script takes raw gene-level counts generated by [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html), scales counts, subsets rows by most significant genes, clusters based by expression similarity across genes and generates a heatmap.  
The output is a heatmap as .png, .jpeg and/or .pdf.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "reshape2",
               "DESeq2",
               "lintr")
for (lib in libraries) {
  if (require(package = lib, character.only = TRUE)) {
    print("Successful")
  } else {
    print("Installing")
    source("https://bioconductor.org/biocLite.R")
    library(lib, character.only = TRUE)
  }
}
## Create output directory:
dir.create(path = "results")
```

Load data:
  
```{r, message = FALSE}
load(file = "results/DESeq2_output_liver.RData")
load(file = "results/eulerr_output_liver.RData")
```

Examine count data:
  
```{r, message = FALSE}
## Normalise counts:
normalised_counts <- t(scale(t(txi$counts),
                             center = TRUE,
                             scale = TRUE))
## Convert to dataframe:
normalised_counts_df <- as.data.frame(normalised_counts)

## Read in sample names and assign to columns of dataframes:
sample_info <- read.table(file = "data/RNAseq_Sample_File_Reduced.txt",
                          header = TRUE)
## Extract liver samples:
sample_info <- subset(x = sample_info,
                      tissue == "Liver")

## Update name for plotting:
## Reorder based on phenotype:
salt_non_smolt_samples  <- c(1, 2, 5, 6, 9, 10, 11, 14)
salt_smolt_samples      <- c(12, 13, 3, 4, 7, 8)
fresh_non_smolt_samples <- c(16, 17, 19, 20, 21, 15, 18, 22)

## Lines appear to be redundant?
#sample_info_ordered <- sample_info[c(1, 2, 5, 6, 9, 10, 11, 14, 12, 13, 3, 4, 7, 8, 16,17,19,20,21,15,18,22),]
#sample_info_ordered

## Reorder the columns by phenotype:
normalised_counts_reordered <- normalised_counts_df[, c(salt_non_smolt_samples,
                                                        salt_smolt_samples,
                                                        fresh_non_smolt_samples)]
## Update sample names:
resident_sw <- "resident-SW"
smolt_sw    <- "smolt-SW"
resident_fw <- "resident-FW"

## Check length of each vector:
#length(grep(pattern = "salt_NonSmolt",
#           colnames(normalised_counts_reordered)))
#length(grep(pattern = "salt_Smolt",
#            colnames(normalised_counts_reordered)))
#length(grep(pattern = "fresh_NonSmolt",
#            colnames(normalised_counts_reordered)))

## Generate new sample names for plotting:
sample_names <- c(paste(resident_sw, "_", "male", "_", 1:5,
                        sep = ""),
                  paste(resident_sw, "_", "female", "_", 1:3,
                        sep = ""),
                  paste(smolt_sw, "_", "male", "_", 1:2,
                        sep = ""),
                  paste(smolt_sw, "_", "female", "_", 1:4,
                        sep = ""),  
                  paste(resident_fw, "_", "male", "_", 1:5,
                        sep = ""),
                  paste(resident_fw, "_", "female", "_", 1:3,
                        sep = ""))

## Rename columns:
colnames(normalised_counts_reordered) <- sample_names
```

Subset genes identified as phenotype but not environment genes:
  
```{r, message = FALSE}
## Check length of vector:
length(env_not_phen_genes)

## Subset genes of interest from dataframe of normalised counts:
normalised_counts_unique_degs <- subset(normalised_counts_reordered,
                                        row.names(normalised_counts_reordered) %in% env_not_phen_genes)
```

Extract normalised counts for genes of interest:
  
```{r, message = FALSE}
## Cluster based on euclidean distance:
dd <- hclust(dist(as.matrix(normalised_counts_unique_degs[, 1:22])))

## Reorder:
normalised_counts_ordered <- as.data.frame(normalised_counts_unique_degs[rev(dd$order), ])
normalised_counts_ordered$locus <- row.names(normalised_counts_ordered)

## Transform dataframe:
normalised_counts_ordered_melt <- melt(normalised_counts_ordered)

## Update column names:
colnames(normalised_counts_ordered_melt) <- c("gene_name",
                                              "sample",
                                              "counts")
## Update levels for plotting:
normalised_counts_ordered_melt$gene_name <-
  factor(normalised_counts_ordered_melt$gene_name,
         levels = rev(unique(c(as.character(unlist(normalised_counts_ordered_melt$gene_name))))))
```

Plot heatmap:
  
```{r, message = FALSE}
## Assign the name of dataframe to plot:
input <- normalised_counts_ordered_melt

## Remove underscores in sample names:
input$sample <- gsub(pattern = "_",
                     replacement = " ",
                     input$sample)

## Update levels for plotting:
input$sample <- factor(input$sample,
                       levels = unique(input$sample))

## Generate heatmap:
heatmap_plot <- ggplot(input,
                       aes(sample,
                           gene_name)) +
        geom_tile(aes(fill = counts)) +
        geom_vline(xintercept = c(0.5,
                                  8.5,
                                  14.5,
                                  22.5)) +
        scale_fill_gradient2(low = "black",
                             mid = "white",
                             high = "red") +
        ylab("") +
        xlab("") +
        theme(legend.title = element_text(size = 10,
                                          face = "bold"),
              legend.text = element_text(size = 10,
                                         face = "bold"),
              axis.text.x = element_text(size = 8,
                                         face = "bold",
                                         colour = "black",
                                         angle = 90,
                                         hjust = 1),
              axis.text.y = element_blank(),
              axis.ticks.y = element_blank(),
              #axis.text.y = element_text(size = 6,
              #                           face = "plain",
              #                           colour = "black"),
              axis.title = element_text(size = 10,
                                        face = "bold"),
              panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              panel.background = element_blank(),
              legend.position = "bottom") +
        scale_y_discrete(position = "right") +
        labs(fill = "Normalised gene expression")

## Print to console:
heatmap_plot

## Save image:
output  <- "results/"
ggsave(file = paste(output,
                    "heatmap_liver_env_not_phen_genes.png",
                    sep = ""),
       height = 6,
       width = 10)

## Save image:
save.image(file = paste(output,
                    "heatmap_liver_env_not_phen_genes.RData",
                    sep = ""))
```

Run lintr:
  
```{r, message = FALSE}
lintr::lint(file = "plot_heatmap_env_not_phen_genes.Rmd")
```


```{r, message = FALSE}
###########################################################
# Next bits are about generating a nice plot with means (as points) and SEs:

#phenotype <- c(rep("resident-SW", 8),
#               rep("smolt-SW", 6),
#               rep("resident-FW", 8))

#input$phenotype <- rep(phenotype,
#                      each = length(env_not_phen_genes))

#sex             <- c(rep("male", 5),
#                     rep("female", 3),
#                     rep("male", 2),
#                     rep("female", 4),
#                     rep("male", 5),
#                     rep("female", 3))

#input$sex <- rep(sex,
#                 each = length(env_not_phen_genes))

## Commands below will add phenotype and sex columns in shorter moves:
input$phenotype <- as.data.frame(str_split_fixed(string = input$sample,
                              pattern = " ",
                              n = 3))$V1
input$sex <- as.data.frame(str_split_fixed(string = input$sample,
                              pattern = " ",
                              n = 3))$V2

# Plot:
par(mfrow = c(1, 3),
    mar   = c(4, 0, 0, 0),
    oma   = c(1, 2, 2, 1))

# Plot mean expression of each gene for residents-SW, splitting by sex:
res_sw_male <- subset(input,
                     phenotype == "resident-SW" &
                     sex == "male")

y <- with(res_sw_male,
          tapply(counts,
                 gene_name,
                 mean))
x <- 1:length(y)

plot(x ~ y,
     pch  = 21,
     col  = "black",
     bg   = "green",
     xlab = "",
     ylab = "Gene index",
     xlim = c(-2, 4))

resSW_female<- subset(input, phenotype=="resident-SW" & sex=="female")
y<- with(resSW_female, tapply(counts, gene_name, mean))
x <- 1:length(y)
points(x~y, pch=1)
mtext("resident-SW")

# Plot mean expression of each gene for smolts-SW, splitting by sex:
smoltSW_male<- subset(input, phenotype=="smolt-SW" & sex=="male")
y<- with(smoltSW_male, tapply(counts, gene_name, mean))
x <- 1:length(y)
plot(x~y, pch=21, col="black", bg="green", xlab="", ylab="Gene index", xlim=c(-2,4))

smoltSW_female<- subset(input, phenotype=="smolt-SW" & sex=="female")
y<- with(smoltSW_female, tapply(counts, gene_name, mean))
x <- 1:length(y)
mtext("smolts-SW")
points(x~y, pch=1)

# Plot mean expression of each gene for residents-FW, splitting by sex:
resFW_male<- subset(input, phenotype=="resident-FW" & sex=="male")
y<- with(resFW_male, tapply(counts, gene_name, mean))
x <- 1:length(y)
plot(x~y, pch=21, col="black", bg="green",  xlab="", ylab="Gene index", xlim=c(-2,4))

resFW_female<- subset(input, phenotype=="resident-FW" & sex=="female")
y<- with(resFW_female, tapply(counts, gene_name, mean))
x <- 1:length(y)
points(x~y, pch=1)
mtext("resident-FW")

################################################################################
## Based on these plots, we can see there are 2 types of genes within these "phen_not_env_genes":

# 1 = The top block of genes are stress genes
# 2 = The bottom block of genes are osmotic environment response genes

# So now let's extract these as 2 separate lists. First, we need to find the break-points between the blocks. The break is most clearly seen in the resident-FW group for females:

y <- with(resFW_female,
          tapply(counts,
                 gene_name,
                 mean))

(y)[1700:2000]

# Elements 1:1751 correspond to the bottom block:
x <- row.names(y)
which(x == "ENSSTUG00000023658.1")

# So genes 1 to 57 in rev(y) correspond to the top block of genes, and we can extract their names as a list like this:
bottom_block_genes_env_not_phen <- x[1:1751]

top_block_genes_env_not_phen <- x[1752:3722]

# Let's look at mean gene expression for each block to make sure the above worked:
resSW <- subset(input,
                phenotype == "resident-SW")
mean(resSW$counts[resSW$gene_name %in% bottom_block_genes_env_not_phen])
mean(resSW$counts[resSW$gene_name %in% top_block_genes_env_not_phen])

smoltsSW <- subset(input,
                   phenotype == "smolt-SW")
mean(smoltsSW$counts[smoltsSW$gene_name %in% bottom_block_genes_env_not_phen])
mean(smoltsSW$counts[smoltsSW$gene_name %in% top_block_genes_env_not_phen])

resFW <- subset(input,
                phenotype == "resident-FW")
mean(resFW$counts[resFW$gene_name %in% bottom_block_genes_env_not_phen])
mean(resFW$counts[resFW$gene_name %in% top_block_genes_env_not_phen])
```

Save as R objects:

```{r, message = FALSE}
## Save a single R object to file:
saveRDS(bottom_block_genes_env_not_phen,
        file = "bottom_block_genes_env_not_phen.rds")

saveRDS(top_block_genes_env_not_phen,
        file = "top_block_genes_env_not_phen.rds")
```


