## Here include some basic information
## Name of author:Robert Wynne
## Date:24/01/2020
## Program name: rscript_rna_seq-1.R
## Purpose of the script:RNAseq analysis
## This script does XX, YYY and ZZZ.
##
## It takes salmon output from ICHEC as input, performs YYY and outputs ZZZ.
## it takes transcription quantification files
## generated by the pseudoaligner salmon, generates
## gene-level counts using Tximport and performs
## differential expression analyses using DESeq2.

##It outputs... - We can work on this when I get back.
## E.g. tables of differentially expressed genes, PCA
## plots.

#RNA seq analysis
```{r, message = FALSE}
# Install libraries:
#install required packages,
#must be done in this way to assure all of the packages work correctly
#if (!requireNamespace("BiocManager", quietly = TRUE))
#install.packages("BiocManager")
#BiocManager::install(version = "3.11")
#BiocManager::install("DESeq2")
#BiocManager::install("apeglm")
#BiocManager::install("tximportData")
#BiocManager::install("DESeqDataSetFromTximport")
#install.packages("ggplot2")
```

Load libraries:

```{r, message = FALSE}
## Load libraries:
library(apeglm)
library(ggplot2)
library(DESeq2)
library(RCurl)
library(stringi)
#library(tximport)
# Check version using sessionInfo
# E.g. sessionInfo("DESeq2")
```

Read in input files. The input files for this script are generated by the pseudoaligner, salmon, which quantifies expression of individual transcript. For differential expression analysis, we want to generate gene-level counts for investigating differences in gene expression between one or more treatment groups.
 
```{r, message = FALSE}
## Check working directory and read in input files:
##  If your script is in the same directory as your input files
## you don't need to set the working directory - remove if redundant.
#set working directory
setwd("~/RNASEQ")
#make sure we are working in the correct directory
getwd()
# read in salmon output files
## JC - This step reads in a list of sample names rather than the actual files
## JC - Don't use capital letters in variable names - 'RNAsamples' should
## be 'rna_samples'
## JC - Actually state what is in 'RNAseq_Sample_File_Reduced.txt' - how many
## columns and what they relate to i.e. First column contains sample names;
## Second column contains the sex of the samples; etc.
rna_sample_info <- read.table(file = "data/RNAseq_Sample_File_Reduced.txt", header = TRUE)
#check the data
## JC - See comment above:
head(rna_sample_info)
## RNASeq analysis for liver samples:
# give R the file path needed, we do this by taking any file from the dirctory with
#'quant.sf attached to the end of the name
## JC - Create a vector containing a list of paths for each sample
## listed in the sample information file read in and stored as 'rna_sample_info'.
## The function 'file.path' will look in the directory 'RNASEQ' and then within indivdiual folders contained within that directory. Each of the individual folders in that directory should contain folders for each sample listed in the sample information file. Within each of these folders, files all end in the file extension 'quant.sf', which is the
## default naming scheme by Salmon.
files <- file.path("data/salmon_output/",
                   rna_sample_info$sample,
                   "quant.sf")
# At present, all samples are being loaded in as input, which contains a mix of liver and brain samples.
#create a dataframe which labels all of the Liver files as true
#this will help is separate the organs as we will be looking at the
#brain and liver separetly
files_df <- as.data.frame(cbind(files,
                                grepl(pattern = "_L",
                                      x = files)))
## creating a subset of just liver samples
## JC: This step subsets the dataframe 'files_df' by checking if values in column 'V2'
## are annotated as 'TRUE'. If so, it subsets these entire lines into a new dataframe
## called 'files_liver_df'.
files_liver_df <- subset(files_df,
                         V2 == TRUE)
## creating a subset with all of the discriptive information for all of the liver samples
##from the 'RNAseqsamples' document
## JC - Again, 'subset' is the function, it takes dataframes as input and creates dataframes
## as output. For this command, the dataframe 'RNAsamples' is subset by the column which contains
## 'tissue' information and only rows related to samples annotated as 'Liver' are subsetted into
## a new dataframe 'RNAsamples_liver'.
rna_sample_info_liver <- subset(rna_sample_info,
                                tissue == "Liver")
## specify the path to the files using the appropriate columns of samples
## also need to link transcripts to genes for this dataset:
tx2gene <- read.table("data/combined_transcript_to_gene.txt")
## Ensure that names are of class character:
files <- as.character(files_liver_df$files)
## Generate gene level counts:
txi <- tximport(files,
                type = "salmon",
                tx2gene = tx2gene)
#check class to make sure data is a list
load("txi.RData")
class(txi)
#save(txi,file = "txi.RData")
### group factors for different comparasons
rna_sample_info_liver$group <- paste(rna_sample_info_liver$environment,
                                     "_",
                                     rna_sample_info_liver$phenotype,
                                     sep = "")


###################################################################################
### For objective 1, we define a global that inclues effects of phenotype (smolt versus resident) and environment (fresh versus salt):
dds_txi_phen_env <- DESeqDataSetFromTximport(txi,
                                   colData = rna_sample_info_liver,
                                   design = ~ phenotype + environment)

### using the results from above to run the DESeq analysis, compare above model with one with only environment. This allows us to get at the significance of the phenotype effect:
dds_env <- DESeq(dds_txi_phen_env,
             test = "LRT",
             reduced = ~environment)

## Check names of comparisons:
resultsNames(dds_env)
## Extract significant results:
phenotype_effect_contr_env <- results(dds,
                      name = "phenotype_Smolt_vs_NonSmolt")

## Extract significant phenotype genes:

phenotype_effect_contr_env$absL2FC <- abs(phenotype_effect_contr_env$log2FoldChange)
phenotype_effect_contr_env_sig <- subset(x = phenotype_effect_contr_env,
                         padj < 0.05 & absL2FC > 1)

## Get gene list:
phenotype_contr_env_sig_genes <- row.names(phenotype_effect_contr_env_sig)
length(phenotype_contr_env_sig_genes)

### using the results from above to run the DESeq analysis, compare above model with one with only phenotype, which allow us to get at the effect of environment:
dds_phen <- DESeq(dds_txi_phen_env,
             test = "LRT",
             reduced = ~phenotype)

## Check names of comparisons:
resultsNames(dds_phen)

## Extract significant results:
env_effect_contr_phen <- results(dds_phen,
                      name = "environment_salt_vs_fresh")
## Extract significant environment genes:

env_effect_contr_phen$absL2FC <- abs(env_effect_contr_phen$log2FoldChange)
env_effect_contr_phen_sig <- subset(x = env_effect_contr_phen,
                         padj < 0.05 & absL2FC > 1)

## Get gene list:
env_contr_phen_sig_genes <- row.names(env_effect_contr_phen_sig)
length(env_contr_phen_sig_genes)

# So we now have two sets of genes:  
# 1 = "phenotype genes" (phenotype_contr_env_sig_genes), i.e. genes that are differentially expressed between smolts and residents, after having controlled statistically for environment
# 2 = "environment genes" (env_contr_phen_sig_genes), i.e. genes that are differentially expressed between fresh and salt water, after having controlled statistically for phenotype.

# We are now in a position to make our venn diagram (see Euler plot code) showing the extent to which these phenotype and environment genes overlap.  But we can see this in numbers like this:
phen_AND_env_genes <- intersect(phenotype_contr_env_sig_genes, env_contr_phen_sig_genes)
length(phen_AND_env_genes)

# So 154 genes overlap between phenotype and environment genes. 

# And here is the number of unique phenotype genes (i.e. DE between phenotypes but not between environments):
phen_NOT_env_genes <- setdiff(phenotype_contr_env_sig_genes, env_contr_phen_sig_genes)
length(phen_NOT_env_genes)

# And here is the number of unique environment genes (i.e. DE between environments but not between phenotypes):
env_NOT_phen_genes <- setdiff(env_contr_phen_sig_genes,phenotype_contr_env_sig_genes)
length(env_NOT_phen_genes)
```


Generate PCA plots to explore expression profiles amongst individuals and treatments:

```{r, message = FALSE}
## JC: Transform data for plotting:
vsd <- vst(dds,
           blind = FALSE)
pca <- prcomp(t(assay(vsd)))
## check the proportion of variance explained by each PC:
summary(pca)
## Here set the two principal components that you want to compare:
## It can any combination between 1 and 22:
first_pc <- 1
second_pc <- 2
## Create function for plotting:
plotPCA.san <- function(object,
                        intgroup = "condition",
                        ntop = 500,
                        returnData = FALSE) 
{
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
                                                     length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/ sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1,
                 paste,
                 collapse = " : "))
  }
  else {
    colData(object)[[intgroup]]
  }
  ## Select the PCAs and percentVar that you like instead of 1 and 2
  d <- data.frame(PC1 = pca$x[ , first_pc],
                  PC2 = pca$x[ , second_pc],
                  group = group, 
                  intgroup.df,
                  name = colData(vsd)[,1])
  if (returnData) {
    attr(d, "percentVar") <- percentVar[first_pc:second_pc]
    return(d)
  }
    ggplot(data = d,
           aes_string(x = "PC1",
                      y = "PC2",
                      color = "group")) +
            geom_point(size = 3) +
            xlab(paste0("PC", first_pc, ": ",
                        round(percentVar[first_pc] * 100),
                        "% variance")) +
            ylab(paste0("PC", second_pc, ": ",
                        round(percentVar[second_pc] * 100),
                        "% variance")) +
            coord_fixed() +
            #geom_text_repel(size = 3) +
            theme_bw() +
            theme(axis.title = element_text(size = 15,
                                            face = "bold"),
                  axis.text = element_text(size = 10,
                                           face = "plain"),
                  legend.title = element_text(size = 10,
                                              face = "bold")) +
            scale_colour_discrete(name = "Phen-env group",
                                labels = c("Resident-FW",
                                           "Resident-SW",
                                           "Smolt-SW"))
}
## Generate plot:
plotPCA.san(vsd,
        intgroup = c("group"))
plotPCA(vsd, intgroup=c("phenotype", "sex"))
plotPCA(vsd, intgroup=c("sex"))
plotPCA(vsd, intgroup=c("phenotype"))
plotPCA(vsd, intgroup=c("environment"))
plotPCA(vsd, intgroup=c("maturity"))
```

Save everything:

```{r, message = FALSE}
save.image(file = "DESeq2_output_liver_tom.RData")
```