--- 
title: "Salmo smoltification study"
output: plot_heatmap.html
author: Joe Colgan (joscolgan)
---

## Introduction:  
This script generates a heatmap for significant differentially expressed genes identified across three experimental groups of brown trout that differ in their phenotype.
The script takes raw gene-level counts generated by [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html), scales counts, subsets rows by most significant genes, clusters based by expression similarity across genes and generates a heatmap.  
The output is a heatmap as .png, .jpeg and/or .pdf.

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "reshape2",
               "DESeq2",
               "lintr")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE)
    }
}

## Create output directory:
dir.create(path = "./results")
```

Load data:

```{r, message = FALSE}
load(file = "../data/DESeq2_output_liver.RData")
```

Examine count data:

```{r, message = FALSE}
## Normalise counts:
normalised_counts <- t(scale(t(txi$counts),
                             center = TRUE,
                             scale = TRUE))

## Convert to dataframe:
normalised_counts_df <- as.data.frame(normalised_counts)

## Read in sample names and assign to columns of dataframes:
sample_info <- read.table(file = "../data/RNAseq_Sample_File_Reduced.txt",
                          header = TRUE)
## Extract liver samples:
sample_info <- subset(x = sample_info,
                      tissue == "Liver")

## Update name for plotting:
colnames(normalised_counts_df) <- paste(sample_info$sample,
                                        "_",
                                        sample_info$environment,
                                        "_",
                                        sample_info$phenotype,
                                        sep = "")

## Reorder based on phenotype:
salt_non_smolt_samples  <- c(1, 2, 5, 6, 9, 10, 11, 14)
salt_smolt_samples      <- c(3, 4, 7, 8, 12, 13)
fresh_non_smolt_samples <- c(15:22)

## Reorder the columns by phenotype:
normalised_counts_reordered <- normalised_counts_df[, c(salt_non_smolt_samples,
                                                        salt_smolt_samples,
                                                        fresh_non_smolt_samples)]
```

Cluster samples by similarity in gene expression profiles:

```{r, message = FALSE}
## First order by adjusted p value to identify the most significant genes:
## Convert to dataframe:
lrt_result_df <- as.data.frame(dds_salt_s_salt_ns)

## Sort by adjusted p value:
lrt_result_df_sorted <- lrt_result_df[order(lrt_result_df$padj), ]

## Extract only significant genes:
lrt_result_df_sorted_sig <- subset(x = lrt_result_df_sorted,
                                   padj < 0.05)

## Assign number of genes to subset:
gene_numbers <- 50

## Subset top fifty most significant genes:
lrt_result_df_sorted_sig <- head(lrt_result_df_sorted_sig,
                                 n = gene_numbers)

## Extract gene ids:
lrt_result_df_sorted_sig_gene_ids <- row.names(lrt_result_df_sorted_sig)
```

Extract normalised counts for genes of interest:

```{r, message = FALSE}
## Subset rows containing gene names of interest:
lrt_result_df_sorted_sig_counts <- normalised_counts_reordered[match(lrt_result_df_sorted_sig_gene_ids,
                                                              row.names(normalised_counts_reordered)), ]
lrt_result_df_sorted_sig_counts$gene_id <- row.names(lrt_result_df_sorted_sig_counts)

## Cluster based on euclidean distance:
dd <- hclust(dist(as.matrix(lrt_result_df_sorted_sig_counts[, 1:22])))

## Reorder:
normalised_counts_ordered <- as.data.frame(lrt_result_df_sorted_sig_counts[rev(dd$order), ])

## Transform dataframe:
normalised_counts_ordered_melt <- melt(normalised_counts_ordered)

## Update column names:
colnames(normalised_counts_ordered_melt) <- c("gene_name",
                                              "sample",
                                              "counts")

## Update levels for plotting:
normalised_counts_ordered_melt$gene_name <-
        factor(normalised_counts_ordered_melt$gene_name,
               levels = rev(unique(c(as.character(unlist(normalised_counts_ordered_melt$gene_name))))))
```

Plot heatmap:

```{r, message = FALSE}
## Assign the name of dataframe to plot:
input <- normalised_counts_ordered_melt

## Remove underscores in sample names:
input$sample <- gsub(pattern = "_",
                     replacement = " ",
                     input$sample)

## Update levels for plotting:
input$sample <- factor(input$sample,
                       levels = unique(input$sample))

## Generate heatmap:
heatmap_plot <- ggplot(input,
                       aes(sample,
                           gene_name)) +
        geom_tile(aes(fill = counts)) +
        geom_vline(xintercept = c(8.5, 14.5, 22.5)) +
        scale_fill_gradient2(low = "black",
                             mid = "white",
                             high = "red") +
        ylab("") +
        xlab("") +
        theme(legend.title = element_text(size = 10,
                                          face = "bold"),
              legend.text = element_text(size = 10,
                                         face = "bold"),
              axis.text.x = element_text(size = 8,
                                         face = "bold",
                                         colour = "black",
                                         angle = 90,
                                         hjust = 1),
              axis.text.y = element_text(size = 6,
                                         face = "plain",
                                         colour = "black"),
              axis.title = element_text(size = 10,
                                        face = "bold"),
              panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              panel.background = element_blank(),
              legend.position = "top") +
        scale_y_discrete(position = "right") +
        labs(fill = "Normalised gene expression")

## Print to console:
heatmap_plot

## Save image:
output  <- "results"
ggsave(file = paste(output,
                    "/",
                    "heatmap_three_groups.png",
                    sep = ""),
       height = 6,
       width = 10)
```

Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./plot_heatmap.Rmd")
```

