
## Here include some basic information
## Name of author:Robert Wynne
## Date:24/01/2020
## Program name: rscript_rna_seq-1.R
## Purpose of the script:RNAseq analysis
## This script does XX, YYY and ZZZ.
##
## It takes salmon output from ICHEC as input, performs YYY and outputs ZZZ.
## it takes transcription quantification files
## generated by the pseudoaligner salmon, generates
## gene-level counts using Tximport and performs
## differential expression analyses using DESeq2.

##It outputs... - We can work on this when I get back.
## E.g. tables of differentially expressed genes, PCA
## plots.

#RNA seq analysis

Load libraries:

```{r, message = FALSE}
## Load libraries:
library(apeglm)
library(ggplot2)
library(DESeq2)
library(RCurl)
library(stringi)
library(tximport)

# Check version using sessionInfo
# E.g. sessionInfo("DESeq2")
```

Read in input files. The input files for this script are generated by the pseudoaligner, salmon, which quantifies expression of individual transcript. For differential expression analysis, we want to generate gene-level counts for investigating differences in gene expression between one or more treatment groups.
 
```{r, message = FALSE}
## Check working directory and read in input files:
##  If your script is in the same directory as your input files
## you don't need to set the working directory - remove if redundant.
#set working directory
setwd("~/")
#make sure we are working in the correct directory
getwd()

## Load R object:
load(file = "../data/DESeq2_output_liver.RData")
```

Create model for investigating interaction term:

```{r, message = FALSE}
### input values and create intermediate calculations and results of an analysis of differential expression looking at sex and
### 'group'(a constructed column combining phenotype and enviornment).
dds_txi <- DESeqDataSetFromTximport(txi,
                                   colData = rna_sample_info_liver,
                                   design = ~sex + group + sex:group)

### using the results from above to run the DESeq analysis
dds <- DESeq(dds_txi,
             test = "LRT",
             reduced = ~sex + group)
```

Question One:  
Is there an interaction effect of group (environment + phenotype) across males in comparison to females that explains gene expression?

```{r, message = FALSE}
## Check names of comparisons:
resultsNames(dds)

## Extract significant results:
lrt_df <- results(dds,
                  name = "sexMale.groupfresh_NonSmolt")

## Extract significant genes:
lrt_sig <- subset(x = lrt_df,
                  padj < 0.05)

## Get gene list:
lrt_sig_genes <- row.names(lrt_sig)
```

Question Two:  
Do sexes differ in expression across the three groups?

```{r, message = FALSE}
## Difference between the sexes:
diff_across_sexes <- results(dds,
                             contrast = c("sex",
                                          "Male",
                                          "Female"))

## Difference between the sexes:
diff_across_sexes_df <- as.data.frame(x = diff_across_sexes)

## Subset significant terms:
diff_across_sexes_sig <- subset(diff_across_sexes_df,
                                padj < 0.05)
```

Compare numbers of differentially expressed genes between treatments of interest:
1) Smolt saltwater vs Non-smolt saltwater
2) Smolt saltwater vs Non-smolt freshwater
3) Non-smolt freshwater vs Non-smolt saltwater

We also want to perform Wald test between comparisons and take genes identified by both approaches as differentially expressed between groups as indentified by the LRT.

```{r, message = FALSE}
### Run a Wald test:
dds_wald <- DESeq(dds_txi,
             test = "Wald")
```

First question: Is there difference in expression between sexes across treatments?

```{r, message = FALSE}
## Check comparison names:
resultsNames(dds_wald)

## Male vs female expression:
dds_sex_male_vs_female <- results(dds_wald,
                                  contrast = c("sex",
                                               "Male",
                                               "Female"))

## Subset:
dds_sex_male_vs_female_df <- as.data.frame(x = dds_sex_male_vs_female)

## Expression between sexes:
dds_sex_male_vs_female_sig <- subset(dds_sex_male_vs_female_df,
                                     padj < 0.05)
```

Second question: Difference in expression between phenotypes?

Extract group comparisons similar to above for the LRTs:
Compare numbers of differentially expressed genes between treatments of interest:
1) Smolt saltwater vs Non-smolt saltwater
2) Smolt saltwater vs Non-smolt freshwater
3) Non-smolt freshwater vs Non-smolt saltwater


```{r, message = FALSE}
## Check comparison names:
resultsNames(dds_wald)

## compare smolts and non smolts
## The effect of the interction between group and sex:
## Comparison 1) Salt non-smolt and fresh non-smolt:
dds_salt_non_smolt_vs_fresh_non_smolt_int <- results(dds_wald,
                                            name = "sexMale.groupsalt_NonSmolt")
## Convert Results output to dataframe:
dds_salt_non_smolt_vs_fresh_non_smolt_int_df <- as.data.frame(dds_salt_non_smolt_vs_fresh_non_smolt_int)
## Subset significant genes:
dds_salt_non_smolt_vs_fresh_non_smolt_int_sig <- subset(dds_salt_non_smolt_vs_fresh_non_smolt_int_df,
                                                    padj < 0.05)

## Comparison 2) Salt smolt and fresh non-smolt:
dds_salt_smolt_vs_fresh_non_smolt_int    <- results(dds_wald,
                                            name = "sexMale.groupsalt_Smolt")
## Convert Results output to dataframe:
dds_salt_smolt_vs_fresh_non_smolt_int_df <- as.data.frame(dds_salt_smolt_vs_fresh_non_smolt_int)
## Subset significant genes:
dds_salt_smolt_vs_fresh_non_smolt_int_sig <- subset(dds_salt_smolt_vs_fresh_non_smolt_int_df,
                                                    padj < 0.05)
```

Reorder reference for Wald test to allow for comparing 

```{r, message = FALSE}
## Relevel the reference group for comparison:
rna_sample_info_liver$group <- relevel(as.factor(rna_sample_info_liver$group),
                                       ref = "salt_Smolt")

### Run a Wald test:
dds_wald <- DESeq(dds_txi,
             test = "Wald")

## Comparison 3) Salt smolt and salt non-smolt:
dds_salt_smolt_vs_salt_non_smolt_int <- results(dds_wald,
                                                name = "sexMale.groupsalt_NonSmolt")
dds_salt_smolt_vs_salt_non_smolt_int_df <- as.data.frame(dds_salt_smolt_vs_salt_non_smolt_int)
dds_salt_smolt_vs_salt_non_smolt_int_sig <- subset(dds_salt_smolt_vs_salt_non_smolt_int_df,
                                                   padj < 0.05)
```

Check overlap between LRT and Wald:

```{r, message = FALSE}
## Extract genes identified as significant by both LRT and Wald:
## Comparison 1) Salt non-smolt and fresh non-smolt:
non_smolt_salt_vs_non_smolts_fresh_lrt_wald <- row.names(subset(dds_salt_smolt_vs_fresh_non_smolt_int_sig,
                                                                row.names(dds_salt_smolt_vs_fresh_non_smolt_int_sig) %in%
                                                                        lrt_sig_genes))
non_smolt_salt_vs_non_smolts_fresh_lrt_wald <- gsub(pattern = "[.]1",
                                                    replacement = "",
                                                    non_smolt_salt_vs_non_smolts_fresh_lrt_wald)
length(non_smolt_salt_vs_non_smolts_fresh_lrt_wald)

## Comparison 2) Salt smolt and fresh non-smolt:
## Extract genes identified as significant by both LRT and Wald:
smolt_vs_non_smolts_fresh_lrt_wald <- row.names(subset(dds_salt_smolt_vs_fresh_non_smolt_int_sig,
                                                       row.names(dds_salt_smolt_vs_fresh_non_smolt_int_sig) %in%
                                                               lrt_sig_genes))
smolt_vs_non_smolts_fresh_lrt_wald <- gsub(pattern = "[.]1",
                                          replacement = "",
                                          smolt_vs_non_smolts_fresh_lrt_wald)
length(smolt_vs_non_smolts_fresh_lrt_wald)

## Comparison 3) Salt smolt and salt non-smolt:
smolt_vs_non_smolts_salt_lrt_wald <- row.names(subset(dds_salt_smolt_vs_salt_non_smolt_int_sig,
                                                      row.names(dds_salt_smolt_vs_salt_non_smolt_int_sig) %in%
                                                              lrt_sig_genes))
smolt_vs_non_smolts_salt_lrt_wald <- gsub(pattern = "[.]1",
                                          replacement = "",
                                          smolt_vs_non_smolts_salt_lrt_wald)
length(smolt_vs_non_smolts_salt_lrt_wald)
```

Subset gene name and p value for all genes analysed:

```{r, message = FALSE}
## For each comparison, extract:
## 1) gene name (taken from the row names of each dataframe) # Required for Euler plots
## 2) unadjusted p value # Required for GO term enrichment
## Comparison 1) Salt non-smolt and fresh non-smolt:
non_smolt_salt_vs_non_smolts_fresh_gene_pvalue <- as.data.frame(cbind(row.names(non_smolt_salt_vs_non_smolts_fresh_df),
                                                            non_smolt_salt_vs_non_smolts_fresh_df[, 5:6]))
## Update column name:
colnames(non_smolt_salt_vs_non_smolts_fresh_gene_pvalue) <- c("locus",
                                                              "raw_pvalue",
                                                              "padj")


smolt_vs_non_smolts_salt_gene_pvalue <- as.data.frame(cbind(row.names(smolt_vs_non_smolts_salt_df),
                                                            smolt_vs_non_smolts_salt_df[, 5:6]))
## Update column name:
colnames(smolt_vs_non_smolts_salt_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")

## Comparison 2) Salt smolt and fresh non-smolt:
smolt_vs_non_smolts_fresh_gene_pvalue <- as.data.frame(cbind(row.names(dds_salt_smolt_vs_fresh_non_smolt_int_df),
                                                            smolt_vs_non_smolts_fresh_df[, 5:6]))
## Update column name:
colnames(smolt_vs_non_smolts_fresh_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")

## Comparison 3: Salt smolt and fresh non-smolt:
smolt_vs_non_smolts_salt_gene_pvalue <- as.data.frame(cbind(row.names(smolt_vs_non_smolts_salt_df),
                                                            smolt_vs_non_smolts_salt_df[, 5:6]))
## Update column name:
colnames(smolt_vs_non_smolts_salt_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")
                                                    
## Comparison 2) Salt smolt and fresh non-smolt:
smolt_vs_non_smolts_fresh_gene_pvalue <- as.data.frame(cbind(row.names(dds_salt_smolt_vs_fresh_non_smolt_int_df),
                                                            smolt_vs_non_smolts_fresh_df[, 5:6]))
## Update column name:
colnames(smolt_vs_non_smolts_fresh_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")
```

Save everything:

```{r, message = FALSE}
save.image(file = "./DESeq2_output_interaction_liver.RData")
```

Run lintr:
```{r, message = FALSE}
require(lintr)
lintr::lint(file = "./deseq2_analysis_interaction.Rmd")
```
