--- 
title: "Salmo smoltification study"
output: deseq2_analysis_sex_interaction.html
authors: Rob Wynne, Joe Colgan (joscolgan)
---

1. Load libraries:

```{r, message = FALSE}
## Load libraries:
library(apeglm)
library(ggplot2)
library(DESeq2)
library(RCurl)
library(stringi)
library(tximport)

# Check version using sessionInfo
# E.g. sessionInfo("DESeq2")
```

2. Read in input files. The input files for this script are generated by the pseudoaligner, salmon, which quantifies expression of individual transcript. For differential expression analysis, we want to generate gene-level counts for investigating differences in gene expression between one or more treatment groups.
 
```{r, message = FALSE}
## Check working directory and read in input files:
##  If your script is in the same directory as your input files
## you don't need to set the working directory - remove if redundant.
#set working directory
setwd("~/")
#make sure we are working in the correct directory
getwd()

## Load R object:
load(file = "results/DESeq2_output_liver.RData")
```

3. Create model for investigating interaction effect between group and sex:

```{r, message = FALSE}
### input values and create intermediate calculations and results of an analysis of differential expression looking at sex and
### 'group'(a constructed column combining phenotype and enviornment).
dds_txi <- DESeqDataSetFromTximport(txi,
                                    colData = rna_sample_info_liver,
                                    design = ~sex + group + sex:group)

### using the results from above to run the DESeq analysis
dds <- DESeq(dds_txi,
             test = "LRT",
             reduced = ~sex + group)
```

4. Investigate interaction across groups + sex:

```{r, message = FALSE}
## Check names of comparisons:
resultsNames(dds)

## Extract significant results:
sex_effect <- results(dds,
                  name = "sex_Male_vs_Female")

## Extract significant genes:
sex_effect_sig <- subset(x = sex_effect,
                         padj < 0.05)

## Extract significant results:
lrt_df <- results(dds,
                  name = "sexMale.groupsalt_NonSmolt")

## Extract significant genes:
lrt_sig <- subset(x = lrt_df,
                  padj < 0.05)

## Get gene list:
lrt_sig_genes <- row.names(lrt_sig)
length(lrt_sig_genes)
```
5. We also want to perform Wald test between comparisons and take significant genes identified by both approaches.

```{r, message = FALSE}
### Run a Wald test:
dds_wald <- DESeq(dds_txi,
             test = "Wald")
```

6. Question: Difference in expression between groups (environment + phenotype)?

Compare numbers of differentially expressed genes between groups to understand the interaction effect of sex on:
1) Smolt saltwater in comparison to Non-smolt freshwater
2) Non-smolt saltwater in comparison to Non-smolt freshwater
3) Smolt saltwater in comparison to Non-smolt saltwater

```{r, message = FALSE}
## Check comparison names:
resultsNames(dds_wald)

## Extract significant results:
sex_effect_wald <- results(dds_wald,
                           name = "sex_Male_vs_Female")

## Extract significant genes:
sex_effect_wald_sig <- subset(x = sex_effect_wald,
                              padj < 0.05)

## compare smolts and non smolts
## The effect of the interction between condition and sex:
## 1) Smolt saltwater in comparison to Non-smolt freshwater:
dds_smolt_SW_vs_resident_FW_int    <- results(dds_wald,
                                              name = "sexMale.groupsalt_Smolt")

## Convert to dataframe:
dds_smolt_SW_vs_resident_FW_int_df <- as.data.frame(dds_smolt_SW_vs_resident_FW_int)
## Subset significant genes:
dds_smolt_SW_vs_resident_FW_int_sig <- subset(dds_smolt_SW_vs_resident_FW_int_df,
                                                    padj < 0.05)
nrow(dds_smolt_SW_vs_resident_FW_int_sig)

## 2) Non-smolt saltwater in comparison to Non-smolt freshwater:
dds_resident_SW_vs_resident_FW_int <- results(dds_wald,
                                                     name = "sexMale.groupsalt_NonSmolt")
## Convert to dataframe:
dds_resident_SW_vs_resident_FW_int_df <- as.data.frame(dds_resident_SW_vs_resident_FW_int)
## Subset significant genes:
dds_resident_SW_vs_resident_FW_int_sig <- subset(dds_resident_SW_vs_resident_FW_int_df,
                                                        padj < 0.05)
nrow(dds_resident_SW_vs_resident_FW_int_sig)
```

7. Reorder reference for Wald test for comparison of:
3) Non-smolt freshwater vs Non-smolt saltwater
At present, the reference level is set to 'fresh_NonSmolt' so can only explore the interaction effect of sex on:

```{r, message = FALSE}
## Relevel the reference group for comparison:
rna_sample_info_liver$group <- relevel(as.factor(rna_sample_info_liver$group),
                                       ref = "salt_NonSmolt")

dds_txi <- DESeqDataSetFromTximport(txi,
                                    colData = rna_sample_info_liver,
                                    design = ~sex + group + sex:group)

### Run a Wald test:
dds_wald <- DESeq(dds_txi,
             test = "Wald")

## 3) Smolt saltwater in comparison to Non-smolt saltwater:
dds_smolt_SW_vs_resident_SW_int <- results(dds_wald,
                                           name = "sexMale.groupsalt_Smolt")

## Convert to dataframe:
dds_smolt_SW_vs_resident_SW_int_df <- as.data.frame(dds_smolt_SW_vs_resident_SW_int)
## Extract significant genes:
dds_smolt_SW_vs_resident_SW_int_sig <- subset(dds_smolt_SW_vs_resident_SW_int_df,
                                              padj < 0.05)
length(dds_smolt_SW_vs_resident_SW_int_sig)
```

8. Check overlap between LRT and Wald:

```{r, message = FALSE}
## Extract genes identified as significant by both LRT and Wald:
## 1) Smolt saltwater in comparison to Non-smolt freshwater:
## Extract genes identified as significant by both LRT and Wald:
dds_smolt_SW_vs_resident_FW_lrt_wald <- row.names(subset(dds_smolt_SW_vs_resident_FW_int_sig,
                                                       row.names(dds_smolt_SW_vs_resident_FW_int_sig) %in%
                                                               lrt_sig_genes))
dds_smolt_SW_vs_resident_FW_lrt_wald <- gsub(pattern = "[.]1",
                                                replacement = "",
                                                dds_smolt_SW_vs_resident_FW_lrt_wald)
length(dds_smolt_SW_vs_resident_FW_lrt_wald)

## 2) Non-smolt saltwater in comparison to Non-smolt freshwater:
## Extract genes identified as significant by both LRT and Wald:
dds_resident_SW_vs_resident_FW_lrt_wald <- row.names(subset(dds_resident_SW_vs_resident_FW_int_sig,
                                                                row.names(dds_resident_SW_vs_resident_FW_int_sig) %in%
                                                                        lrt_sig_genes))
dds_resident_SW_vs_resident_FW_lrt_wald <- gsub(pattern = "[.]1",
                                                replacement = "",
                                                dds_resident_SW_vs_resident_FW_lrt_wald)
length(dds_resident_SW_vs_resident_FW_lrt_wald)

## 3) Smolt saltwater in comparison to Non-smolt saltwater:
dds_smolt_SW_vs_resident_SW_lrt_wald <- row.names(subset(dds_smolt_SW_vs_resident_SW_int_sig,
                                                      row.names(dds_smolt_SW_vs_resident_SW_int_sig) %in%
                                                              lrt_sig_genes))
dds_smolt_SW_vs_resident_SW_lrt_wald <- gsub(pattern = "[.]1",
                                             replacement = "",
                                             dds_smolt_SW_vs_resident_SW_lrt_wald)
length(dds_smolt_SW_vs_resident_SW_lrt_wald)
```

9. Subset gene name and p value for all genes analysed:

```{r, message = FALSE}
## For each comparison, extract:
## 1) gene name (taken from the row names of each dataframe) # Required for Euler plots
## 2) unadjusted p value # Required for GO term enrichment
## 1) Smolt saltwater in comparison to Non-smolt freshwater:
dds_smolt_SW_vs_resident_FW_gene_pvalue <- as.data.frame(cbind(row.names(dds_smolt_SW_vs_resident_FW_int_df),
                                                            dds_smolt_SW_vs_resident_FW_int_df[, 5:6]))
## Update column names:
colnames(dds_smolt_SW_vs_resident_FW_gene_pvalue) <- c("locus",
                                                       "raw_pvalue",
                                                       "padj")

## 2) Non-smolt saltwater in comparison to Non-smolt freshwater:
dds_resident_SW_vs_resident_FW_gene_pvalue <- as.data.frame(cbind(row.names(dds_resident_SW_vs_resident_FW_int_df),
                                                            dds_resident_SW_vs_resident_FW_int_df[, 5:6]))
## Update column names:
colnames(dds_resident_SW_vs_resident_FW_gene_pvalue) <- c("locus",
                                                          "raw_pvalue",
                                                          "padj")

## 3) Smolt saltwater in comparison to Non-smolt saltwater:
dds_smolt_SW_vs_resident_SW_gene_pvalue <- as.data.frame(cbind(row.names(dds_smolt_SW_vs_resident_SW_int_df),
                                                            dds_smolt_SW_vs_resident_SW_int_df[, 5:6]))
## Update column names:
colnames(dds_smolt_SW_vs_resident_SW_gene_pvalue) <- c("locus",
                                                       "raw_pvalue",
                                                       "padj")
```

10. Check overlap among comparisons:

```{r, message = FALSE}
## check number of overlapping genes across all comparisons:
## A) Are there genes where a significant interaction effect between sex and group is identified for:
## smolt in comparison to resident freshwater and;
## resident saltwater in comparison to resident freshwater and;
## smolt in comparison to resident saltwater?
table(table(c(dds_smolt_SW_vs_resident_FW_lrt_wald,
              dds_resident_SW_vs_resident_FW_lrt_wald,
              dds_smolt_SW_vs_resident_SW_lrt_wald)))

## check number of overlapping genes across each comparison:
## B) Are there genes where a significant interaction effect between sex and group is identified for:
## smolt in comparison to resident freshwater and;
## resident saltwater in comparison to resident freshwater?
table(table(c(dds_smolt_SW_vs_resident_FW_lrt_wald,
              dds_resident_SW_vs_resident_FW_lrt_wald)))

## check number of overlapping genes across each comparison:
## C) Are there genes where a significant interaction effect between sex and group is identified for:
## smolt in comparison to resident freshwater and;
## smolt in comparison to resident saltwater?
table(table(c(dds_smolt_SW_vs_resident_FW_lrt_wald,
              dds_smolt_SW_vs_resident_SW_lrt_wald)))

## check number of overlapping genes across each comparison:
## D) Are there genes where a significant interaction effect between sex and group is identified for:
## resident salwater in comparison to resident freshwater;
## smolt in comparison to resident saltwater?
table(table(c(dds_resident_SW_vs_resident_FW_lrt_wald,
              dds_smolt_SW_vs_resident_SW_lrt_wald)))
```

11. Check overlap with genes that are differentially expressed between sexes (i.e. main effect of sex on gene expression).

```{r, message = FALSE}
## Overlap with the main effect of sex on gene expression:
sex_effect_wald_gene_list <- row.names(sex_effect_wald_sig)
sex_effect_wald_gene_list <- gsub(pattern = "[.]1",
                                  replacement = "",
                                  sex_effect_wald_gene_list)

## Overlap between genes differentially expressed by sex and:
## a) smolt SW vs resident FW: 
table(table(c(dds_smolt_SW_vs_resident_FW_lrt_wald,
              sex_effect_wald_gene_list)))

## b) resident SW vs resident FW: 
table(table(c(dds_resident_SW_vs_resident_FW_lrt_wald,
              sex_effect_wald_gene_list)))

## c) smolt SW vs resident SW:
table(table(c(dds_smolt_SW_vs_resident_SW_lrt_wald,
              sex_effect_wald_gene_list)))
```

12. Save everything:

```{r, message = FALSE}
save.image(file = "DESeq2_output_interaction_liver.RData")
```
