
## Here include some basic information
## Name of author:Robert Wynne
## Date:24/01/2020
## Program name: rscript_rna_seq-1.R
## Purpose of the script:RNAseq analysis
## This script does XX, YYY and ZZZ.
##
## It takes salmon output from ICHEC as input, performs YYY and outputs ZZZ.
## it takes transcription quantification files
## generated by the pseudoaligner salmon, generates
## gene-level counts using Tximport and performs
## differential expression analyses using DESeq2.

##It outputs... - We can work on this when I get back.
## E.g. tables of differentially expressed genes, PCA
## plots.

#RNA seq analysis

Load libraries:

```{r, message = FALSE}
## Load libraries:
library(apeglm)
library(ggplot2)
library(DESeq2)
library(RCurl)
library(stringi)
library(tximport)

# Check version using sessionInfo
# E.g. sessionInfo("DESeq2")
```

Read in input files. The input files for this script are generated by the pseudoaligner, salmon, which quantifies expression of individual transcript. For differential expression analysis, we want to generate gene-level counts for investigating differences in gene expression between one or more treatment groups.
 
```{r, message = FALSE}
## Check working directory and read in input files:
##  If your script is in the same directory as your input files
## you don't need to set the working directory - remove if redundant.
#set working directory
setwd("~/")
#make sure we are working in the correct directory
getwd()

## Load R object:
load(file = "results/DESeq2_output_liver.RData")
```

Create model for investigating interaction effect between group and sex:

```{r, message = FALSE}
### input values and create intermediate calculations and results of an analysis of differential expression looking at sex and
### 'group'(a constructed column combining phenotype and enviornment).
dds_txi <- DESeqDataSetFromTximport(txi,
                                    colData = rna_sample_info_liver,
                                    design = ~sex + group + sex:group)

### using the results from above to run the DESeq analysis
dds <- DESeq(dds_txi,
             test = "LRT",
             reduced = ~sex + group)
```

Investigate interaction across groups + sex:

```{r, message = FALSE}
## Check names of comparisons:
resultsNames(dds)

## Extract significant results:
lrt_df <- results(dds,
                  name = "sexMale.groupsalt_NonSmolt")

## Extract significant genes:
lrt_sig <- subset(x = lrt_df,
                  padj < 0.05)

## Get gene list:
lrt_sig_genes <- row.names(lrt_sig)
length(lrt_sig_genes)
```
We also want to perform Wald test between comparisons and take significant genes identified by both approaches.

```{r, message = FALSE}
### Run a Wald test:
dds_wald <- DESeq(dds_txi,
             test = "Wald")
```

Question: Difference in expression between groups (environment + phenotype)?

Compare numbers of differentially expressed genes between groups to understand the interaction effect of sex on:
1) Smolt saltwater in comparison to Non-smolt freshwater
2) Non-smolt saltwater in comparison to Non-smolt freshwater
3) Smolt saltwater in comparison to Non-smolt saltwater

```{r, message = FALSE}
## Check comparison names:
resultsNames(dds_wald)

## compare smolts and non smolts
## The effect of the interction between condition and sex:
## 1) Smolt saltwater in comparison to Non-smolt freshwater:
dds_salt_smolt_vs_fresh_non_smolt_int    <- results(dds_wald,
                                                    name = "sexMale.groupsalt_Smolt")

## Convert to dataframe:
dds_salt_smolt_vs_fresh_non_smolt_int_df <- as.data.frame(dds_salt_smolt_vs_fresh_non_smolt_int)
## Subset significant genes:
dds_salt_smolt_vs_fresh_non_smolt_int_sig <- subset(dds_salt_smolt_vs_fresh_non_smolt_int_df,
                                                    padj < 0.05)
nrow(dds_salt_smolt_vs_fresh_non_smolt_int_sig)

## 2) Non-smolt saltwater in comparison to Non-smolt freshwater:
dds_salt_non_smolt_vs_fresh_non_smolt_int <- results(dds_wald,
                                                     name = "sexMale.groupsalt_NonSmolt")
## Convert to dataframe:
dds_salt_non_smolt_vs_fresh_non_smolt_int_df <- as.data.frame(dds_salt_non_smolt_vs_fresh_non_smolt_int)
## Subset significant genes:
dds_salt_non_smolt_vs_fresh_non_smolt_int_sig <- subset(dds_salt_non_smolt_vs_fresh_non_smolt_int_df,
                                                        padj < 0.05)
nrow(dds_salt_non_smolt_vs_fresh_non_smolt_int_sig)
```

Reorder reference for Wald test for comparison of:
3) Non-smolt freshwater vs Non-smolt saltwater
At present, the reference level is set to 'fresh_NonSmolt' so can only explore the interaction effect of sex on:

```{r, message = FALSE}
## Relevel the reference group for comparison:
rna_sample_info_liver$group <- relevel(as.factor(rna_sample_info_liver$group),
                                       ref = "salt_NonSmolt")

dds_txi <- DESeqDataSetFromTximport(txi,
                                    colData = rna_sample_info_liver,
                                    design = ~sex + group + sex:group)

### Run a Wald test:
dds_wald <- DESeq(dds_txi,
             test = "Wald")

## 3) Smolt saltwater in comparison to Non-smolt saltwater:
dds_salt_smolt_vs_salt_non_smolt_int <- results(dds_wald,
                                                 name = "sexMale.groupsalt_Smolt")

## Convert to dataframe:
dds_salt_smolt_vs_salt_non_smolt_int_df <- as.data.frame(dds_salt_smolt_vs_salt_non_smolt_int)
## Extract significant genes:
dds_salt_smolt_vs_salt_non_smolt_int_sig <- subset(dds_salt_smolt_vs_salt_non_smolt_int_df,
                                                    padj < 0.05)
length(dds_salt_smolt_vs_salt_non_smolt_int_sig)
```

Check overlap between LRT and Wald:

```{r, message = FALSE}
## Extract genes identified as significant by both LRT and Wald:
## 1) Smolt saltwater in comparison to Non-smolt freshwater:
## Extract genes identified as significant by both LRT and Wald:
smolt_vs_non_smolts_fresh_lrt_wald <- row.names(subset(dds_salt_smolt_vs_fresh_non_smolt_int_sig,
                                                       row.names(dds_salt_smolt_vs_fresh_non_smolt_int_sig) %in%
                                                               lrt_sig_genes))
smolt_vs_non_smolts_fresh_lrt_wald <- gsub(pattern = "[.]1",
                                          replacement = "",
                                          smolt_vs_non_smolts_fresh_lrt_wald)
length(smolt_vs_non_smolts_fresh_lrt_wald)

## 2) Non-smolt saltwater in comparison to Non-smolt freshwater:
## Extract genes identified as significant by both LRT and Wald:
non_smolt_salt_vs_non_smolts_fresh_lrt_wald <- row.names(subset(dds_salt_non_smolt_vs_fresh_non_smolt_int_sig,
                                                                row.names(dds_salt_non_smolt_vs_fresh_non_smolt_int_sig) %in%
                                                                        lrt_sig_genes))
non_smolt_salt_vs_non_smolts_fresh_lrt_wald <- gsub(pattern = "[.]1",
                                                    replacement = "",
                                                    non_smolt_salt_vs_non_smolts_fresh_lrt_wald)
length(non_smolt_salt_vs_non_smolts_fresh_lrt_wald)

## 3) Smolt saltwater in comparison to Non-smolt saltwater:
smolt_vs_non_smolts_salt_lrt_wald <- row.names(subset(dds_salt_smolt_vs_salt_non_smolt_int_sig,
                                                      row.names(dds_salt_smolt_vs_salt_non_smolt_int_sig) %in%
                                                              lrt_sig_genes))
smolt_vs_non_smolts_salt_lrt_wald <- gsub(pattern = "[.]1",
                                          replacement = "",
                                          smolt_vs_non_smolts_salt_lrt_wald)
length(smolt_vs_non_smolts_salt_lrt_wald)
```

Subset gene name and p value for all genes analysed:

```{r, message = FALSE}
## For each comparison, extract:
## 1) gene name (taken from the row names of each dataframe) # Required for Euler plots
## 2) unadjusted p value # Required for GO term enrichment
## 1) Smolt saltwater in comparison to Non-smolt freshwater:
smolt_vs_non_smolts_fresh_gene_pvalue <- as.data.frame(cbind(row.names(dds_salt_smolt_vs_fresh_non_smolt_int_df),
                                                            smolt_vs_non_smolts_fresh_df[, 5:6]))
## Update column names:
colnames(smolt_vs_non_smolts_fresh_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")

## 2) Non-smolt saltwater in comparison to Non-smolt freshwater:
non_smolt_salt_vs_non_smolts_fresh_gene_pvalue <- as.data.frame(cbind(row.names(non_smolt_salt_vs_non_smolts_fresh_df),
                                                            non_smolt_salt_vs_non_smolts_fresh_df[, 5:6]))
## Update column names:
colnames(non_smolt_salt_vs_non_smolts_fresh_gene_pvalue) <- c("locus",
                                                              "raw_pvalue",
                                                              "padj")

## 3) Smolt saltwater in comparison to Non-smolt saltwater:
smolt_vs_non_smolts_salt_gene_pvalue <- as.data.frame(cbind(row.names(smolt_vs_non_smolts_salt_df),
                                                            smolt_vs_non_smolts_salt_df[, 5:6]))
## Update column names:
colnames(smolt_vs_non_smolts_salt_gene_pvalue) <- c("locus",
                                                    "raw_pvalue",
                                                    "padj")
```

Save everything:

```{r, message = FALSE}
save.image(file = "DESeq2_output_interaction_liver.RData")
```
