--- 
title: "Salmo sex study"
output: plot_distribution.html
author: Joe Colgan (joscolgan)
---

## Introduction:

1. Load libraries for performing analysis:

```{r, message = FALSE}
libraries <- c("devtools",
               "ggplot2",
               "PopGenome",
               "qqman",
               "reshape",
               "ggpubr",
               "zoo",
               "stringr",
               "dplyr",
               "ggrepel")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                chbiocLite(pkgs = lib)
                library(lib, character.only = TRUE)
        }
}
```

Load input data:

```{r, message = FALSE}
load(file = "DESeq2_output_liver.RData")
```

Load in file containing chromosome information:

```{r, message = FALSE}
gene_location_info <- read.table(file = "data/ensembl/gff/Salmo_trutta.fSalTru1.1.100.genes.bed",
                                 header = FALSE,
                                 col.names = c("chrom",
                                               "start",
                                               "end",
                                               "locus"))

## Convert "chrom" to numeric value:
gene_location_info$chrom <- as.numeric(unlist(gene_location_info$chrom))
#gene_location_info$locus <- as.character(unlist(gene_location_info$locus))

## Subset to retain only genes on chromosomes:
#gene_location_info <- subset(gene_location_info,
#                             chrom <= 40)
```



```{r, message = FALSE}
## Reformat gene names for subsetting:
row.names(smolt_vs_non_smolts_fresh_df) <- gsub(pattern = "[.]1",
                                        replacement = "",
                                        row.names(smolt_vs_non_smolts_fresh_df))

## Add chromosome information:
#smolt_vs_non_smolts_fresh_df$chrom <- gene_location_info[match(smolt_vs_non_smolts_fresh_genes,
#                                                                      gene_location_info$locus), ]$chrom

## Subset gene coordinates:
tmp <- as.data.frame(gene_location_info[match(row.names(smolt_vs_non_smolts_fresh_df),
                                                                      gene_location_info$locus), ])

smolt_vs_non_smolts_fresh_df <- cbind(smolt_vs_non_smolts_fresh_df,
                                      tmp)

smolt_vs_non_smolts_fresh_df <- subset(smolt_vs_non_smolts_fresh_df,
                                         chrom <= 40)

## columns required for Manhattan plot:
input_for_manhattan <- smolt_vs_non_smolts_fresh_df %>% select(chrom,
                                                               locus,
                                                               start,
                                                               log2FoldChange)

## Filter based on chromosome:
combined_filtered <- subset(input_for_manhattan,
                            as.numeric(chrom) <= 40)

## Sort by chromosome and position:
combined_sorted <- combined_filtered[
  with(combined_filtered, order(chrom,
                                start)),
]

## Subset genes that have significantly high and low expression:
smolt_vs_non_smolts_fresh_high <- subset(smolt_vs_non_smolts_fresh_df,
                                        log2FoldChange > 1 &
                                        padj < 0.01)
smolt_vs_non_smolts_fresh_low  <- subset(smolt_vs_non_smolts_fresh_df,
                                        log2FoldChange < -1 &
                                        padj < 0.01)
## Combine:
smolt_vs_non_smolts_fresh_sig  <- rbind(smolt_vs_non_smolts_fresh_high,
                                        smolt_vs_non_smolts_fresh_low)

smolt_vs_non_smolts_fresh_subset <- subset(smolt_vs_non_smolts_fresh_sig,
       !(row.names(smolt_vs_non_smolts_fresh_sig) %in%
                 c(row.names(smolt_vs_non_smolts_salt_sig),
                   row.names(non_smolt_salt_vs_non_smolts_fresh_sig))))

sig_genes <- smolt_vs_non_smolts_fresh_subset$locus

## Convert CHR to factor and sort levels by correct order:
combined_sorted$chrom <- factor(combined_sorted$chrom,
                               levels = unique(as.numeric(combined_sorted$chrom)))
options(scipen = 999)

class(combined_sorted$start)

don <- combined_sorted %>%
  # Compute chromosome size
  group_by(chrom) %>%
  summarise(chr_len = max(start)) %>%
  # Calculate cumulative position of each chromosome
  mutate(tot = cumsum(as.numeric(chr_len)) - chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(combined_sorted, ., by = c("chrom" = "chrom")) %>%
  # Add a cumulative position of each SNP
  arrange(chrom, start) %>%
  mutate(BPcum = start + tot) %>%
  # Add highlight and annotation information
  mutate(is_highlight = ifelse(locus %in% sig_genes,
                                "yes",
                                "no"))

axisdf <- don %>% group_by(chrom) %>%
          summarize(center = (max(BPcum) + min(BPcum)) / 2)

# Make the plot
manhattan_plot <- ggplot(don, aes(x = BPcum,
                                  y = log2FoldChange)) +
    # Show all points
    geom_point(aes(color = as.factor(chrom)),
                alpha = 0.3,
                size = .3) +
    scale_color_manual(values = rep(c("blue",
                                      "black"),
                                    22)) +
    xlab("Chromosome") +
    ylab("Log2FoldChange") +
    # custom X axis:
    scale_x_continuous(label = axisdf$chrom,
                       breaks = axisdf$center) +
    scale_y_continuous(expand = c(0.01, 0.01)) +
    ylim(c(-10, 10)) +
    # Add highlighted points
    geom_point(data = subset(don,
                             is_highlight == "yes"),
               color = "darkorange",
               alpha = 0.8,
               size = 1.3) +
    # Custom the theme:
    theme_bw() +
    theme(
      legend.position = "none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text = element_text(size = 10,
                               face = "plain"),
      axis.title = element_text(size = 15,
                              face = "bold"))
```



```{r, message = FALSE}
## Reformat gene names for subsetting:
smolt_vs_non_smolts_fresh_genes <- gsub(pattern = "[.]1",
                                        replacement = "",
                                        row.names(smolt_vs_non_smolts_fresh_sig))

## Match and subset genes of interest:
smolt_vs_non_smolts_fresh_subset <- subset(smolt_vs_non_smolts_fresh_sig,
       !(row.names(smolt_vs_non_smolts_fresh_sig) %in%
                 c(row.names(smolt_vs_non_smolts_salt_sig),
                   row.names(non_smolt_salt_vs_non_smolts_fresh_sig))))

## Add chromosome information:
smolt_vs_non_smolts_fresh_sig$chrom <- as.numeric(unlist(gene_location_info[match(smolt_vs_non_smolts_fresh_genes,
                                                                      gene_location_info$locus), ]$chrom))
```


```{r, message = FALSE}
## Read in file containing output from DESeq2
degs <- read.table(file = "./input/differentially_expressed_genes.txt",
                   header = TRUE)

## Create a vector containing gene names:
full_genes <- rownames(degs)
full_genes <- gsub(pattern = "[.]1",
                   replacement = "",
                   full_genes)

## Subset significant genes:
sig_genes <- row.names(subset(degs,
                              padj < 0.05))

## Subset count dataframe for genes where deg analysis was performed:
restricted <- subset(counts_df,
                     rownames(counts_df) %in% full_genes)

gene_info_reduced <- subset(gene_info,
                            gene_name %in% full_genes)

## Order degs based on significance:
degs_order <- degs[order(degs$padj), ]

## Extract gene names of most significant genes:
most_sig_order <- rownames(degs_order)

nrow(degs)
row.names(degs) <- gsub(pattern = "[.]1",
                        replacement = "",
                        row.names(degs))

nrow(gene_info_reduced)

degs_reduced <- subset(degs,
                       row.names(degs) %in% gene_info_reduced$gene_name)

## Need to sort both by gene name:
degs_reduced_ordered <- degs_reduced[order(row.names(degs_reduced)), ]
gene_info_ordered <- gene_info_reduced[order(gene_info_reduced$gene_name), ]

combined_df <- as.data.frame(cbind(degs_reduced_ordered,
                                   gene_info_ordered))

## Filter based on chromosome:
combined_filtered <- subset(combined_df,
                            as.numeric(chromosome) <= 40)

## Sort by chromosome and position:
combined_sorted <- combined_filtered[
  with(combined_filtered, order(chromosome,
                                position)),
]
```



```{r, message = FALSE}
## Convert CHR to factor and sort levels by correct order:
combined_sorted$chromosome <- factor(combined_sorted$chromosome,
                               levels = unique(as.numeric(combined_sorted$chromosome)))
options(scipen = 999)

class(combined_sorted$position)

don <- combined_sorted %>%
  # Compute chromosome size
  group_by(chromosome) %>%
  summarise(chr_len = max(position)) %>%
  # Calculate cumulative position of each chromosome
  mutate(tot = cumsum(as.numeric(chr_len)) - chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(combined_sorted, ., by = c("chromosome" = "chromosome")) %>%
  # Add a cumulative position of each SNP
  arrange(chromosome, position) %>%
  mutate(BPcum = position + tot) %>%
  # Add highlight and annotation information
  mutate(is_highlight = ifelse(gene_name %in% sig_genes,
                                "yes",
                                "no"))

axisdf <- don %>% group_by(chromosome) %>%
          summarize(center = (max(BPcum) + min(BPcum)) / 2)

don$log <- log10(don$padj)

# Make the plot
manhattan_plot <- ggplot(don, aes(x = BPcum,
                                  y = log)) +
    # Show all points
    geom_point(aes(color = as.factor(chromosome)),
                alpha = 0.3,
                size = .3) +
    scale_color_manual(values = rep(c("black",
                                      "navy"),
                                    22)) +
    geom_hline(yintercept = 2.56,
               color = "black",
               linetype = "dashed") +
    xlab("Chromosome") +
    ylab("-log10(p)") +
    # custom X axis:
    scale_x_continuous(label = axisdf$chromosome,
                       breaks = axisdf$center) +
    scale_y_continuous(expand = c(0.01, 0.01)) +
    # Add highlighted points
    geom_point(data = subset(don,
                             is_highlight == "yes"),
               color = "darkorange",
               alpha = 0.8,
               size = 1.3) +
    # Add label using ggrepel to avoid overlapping
    #geom_label_repel( data = subset(don,
    #                                is_highlight == "yes"),
    #                  aes(label = locus),
    #                  size = 2.5) +
    # Custom the theme:
    theme_bw() +
    theme(
      legend.position = "none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text = element_text(size = 10,
                               face = "plain"),
      axis.title = element_text(size = 15,
                              face = "bold"))

## Save to file:
dir.create(path = "./results/figures",
           recursive = TRUE)
ggsave(file = "./results/figures/manhattan_plot.png",
       width = 15,
       height = 8)

log_manhattan_plot <- ggplot(don, aes(x = BPcum,
                                  y = log2FoldChange)) +
    # Show all points
    geom_point(aes(color = as.factor(chromosome)),
                alpha = 0.3,
                size = .3) +
    scale_color_manual(values = rep(c("black",
                                      "navy"),
                                    22)) +
    xlab("Chromosome") +
    ylab("logFC") +
    # custom X axis:
    scale_x_continuous(label = axisdf$chromosome,
                       breaks = axisdf$center) +
    scale_y_continuous(expand = c(0.01, 0.01)) +
    # Add highlighted points
    geom_point(data = subset(don,
                             is_highlight == "yes"),
               color = "darkorange",
               alpha = 0.5,
               size = 1.3) +
    # Add label using ggrepel to avoid overlapping
    #geom_label_repel( data = subset(don,
    #                                is_highlight == "yes"),
    #                  aes(label = locus),
    #                  size = 2.5) +
    # Custom the theme:
    theme_bw() +
    theme(
      legend.position = "none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text = element_text(size = 10,
                               face = "plain"),
      axis.title = element_text(size = 15,
                              face = "bold"))

## Save to file:
ggsave(file = "./results/figures/log_manhattan_plot.png",
       width = 15,
       height = 8)
```

Explore the presence of sex-limited expression:

```{r, message = FALSE}
nrow(counts_df)

## Subset male samples:
female_names <- grep("_F_",
     colnames(counts_df))
female_names_df <- counts_df[, female_names]
female_names_df$sum <- rowSums(female_names_df)
## Subset zero genes:
female_names_zero <- subset(female_names_df,
                          sum == 0)
female_names_zero_genes <- rownames(female_names_zero)
male_names <- grep("_M_",
     colnames(counts_df))
male_names_df <- counts_df[, male_names]
male_names_df$sum <- rowSums(male_names_df)
## Subset zero genes:
male_names_zero <- subset(male_names_df,
                          sum == 0)
male_names_zero_genes <- rownames(male_names_zero)

##
female_names_male_zeros <- subset(female_names_df,
       rownames(female_names_df) %in% male_names_zero_genes)

test <- female_names_male_zeros[apply(female_names_male_zeros[, -1],
                                      MARGIN = 1,
                                      function(x) all(x > 10)), ]

##
male_names_female_zeros <- subset(male_names_df,
       rownames(male_names_df) %in% female_names_zero_genes)

test_2 <- male_names_female_zeros[apply(male_names_female_zeros[, -1],
                                        MARGIN = 1,
                                        function(x) all(x > 10)), ]
```

Run lintr:

```{r, message = FALSE}
lintr::lint(file = "./plot_distribution.Rmd")
```

