--- 
  title: "Salmo smoltification study"
output: deseq2_analysis_sex_interaction.html
authors: Rob Wynne, Joe Colgan (joscolgan)
---
  
  1. Load libraries:
  
```{r, message = FALSE}
## Load libraries:
library(apeglm)
library(ggplot2)
library(DESeq2)
library(RCurl)
library(stringi)
library(tximport)
# Check version using sessionInfo
# E.g. sessionInfo("DESeq2")
```

2. Read in input files. The input files for this script are generated by the pseudoaligner, salmon, which quantifies expression of individual transcript. For differential expression analysis, we want to generate gene-level counts for investigating differences in gene expression between one or more treatment groups.

```{r, message = FALSE}
## Check working directory and read in input files:
##  If your script is in the same directory as your input files
## you don't need to set the working directory - remove if redundant.
#set working directory
setwd("~/")
#make sure we are working in the correct directory
getwd()
## Load R object:
load(file = "DESeq2_output_brain_tom.RData")
```

3. Create model for investigating interaction effect between group and sex:
  
```{r, message = FALSE}
### input values and create intermediate calculations and results of an analysis of differential expression looking at sex and
### 'group'(a constructed column combining phenotype and enviornment).
dds_sex <- DESeqDataSetFromTximport(txi,
                                    colData = rna_sample_info_liver,
                                    design = ~sex)
### using the results from above to run the DESeq analysis
dds_null <- DESeq(dds_sex,
             test = "LRT",
             reduced = ~ 1)
```

4. Investigate main effect of sex:
  
```{r, message = FALSE}
## Check names of comparisons:
resultsNames(dds_null)
## Extract significant results:
sex_effect <- results(dds_null,
                      name = "sex_Male_vs_Female")

sex_effect$absL2FC <- abs(sex_effect$log2FoldChange)
## Extract significant genes:
sex_effect_sig <- subset(x = sex_effect,
                         padj < 0.05 & absL2FC > 1)

## Get gene list:
sex_sig_genes <- row.names(sex_effect_sig)
length(sex_sig_genes)

###############################################################################
## Overlap between main effect of sex and ALH genes:
ALH_sex_effect_genes<- intersect(sex_sig_genes, phen_not_env_genes)
length(ALH_sex_effect_genes)
length(ALH_sex_effect_genes)/length(phen_not_env_genes)

stress_sex_effect_genes <- intersect(sex_sig_genes, phen_and_env_genes)
length(stress_sex_effect_genes)
length(stress_sex_effect_genes)/length(phen_and_env_genes)

osm_env_response_sex_effect_genes<- intersect(sex_sig_genes, env_not_phen_genes)
length(osm_env_response_sex_effect_genes)
length(osm_env_response_sex_effect_genes)/length(env_not_phen_genes)


nonALH_sex_effect_genes <- setdiff(sex_sig_genes, ALH_genes)
length(nonALH_sex_effect_genes)

all_genes <- row.names(lrt_raw_df)
length(all_genes)
length(nonALH_sex_effect_genes)/(length(all_genes)-length(ALH_genes))

# Let's now make a contingency table of ALH/non-ALH genes versus sex-biased/unbiased genes:
x <- as.table(rbind(c(length(ALH_sex_effect_genes), length(nonALH_sex_effect_genes)), c((length(ALH_genes)-length(ALH_sex_effect_genes)), (length(all_genes)-length(ALH_genes)-length(nonALH_sex_effect_genes)))))

x
row.names(x)<- c("Sex-biased", "Unbiased")
colnames(x)<- c("ALH genes", "non-ALH genes")
x

# The total needs to sum to length(all_genes):
sum(x)
length(all_genes)

# And now a chi square test:
chisq.test(x)

# The above shows that there is a statistically significant excess of sex-biased ALH genes in comparison to sex-biased non-ALH genes. 

###############
# Now let's make a contingency table of stress/non-stress genes versus sex-biased/unbiased genes:
nonstress_sex_effect_genes <- setdiff(sex_sig_genes, stress_genes)
length(nonstress_sex_effect_genes)

x <- as.table(rbind(c(length(stress_sex_effect_genes), length(nonstress_sex_effect_genes)), c((length(stress_genes)-length(stress_sex_effect_genes)), (length(all_genes)-length(stress_genes)-length(nonstress_sex_effect_genes)))))

x
row.names(x)<- c("Sex-biased", "Unbiased")
colnames(x)<- c("stress genes", "non-stress genes")
x

# The total needs to sum to length(all_genes):
sum(x)
length(all_genes)

# And now a chi square test:
chisq.test(x)

# The above shows that there are no more sex-biased stress genes than you'd expect by chance.

###############
# Now let's make a contingency table of stress/non-stress genes versus sex-biased/unbiased genes:
non_osm_env_sex_effect_genes <- setdiff(sex_sig_genes, osm_env_response_genes)
length(non_osm_env_sex_effect_genes)

x <- as.table(rbind(c(length(osm_env_response_sex_effect_genes), length(non_osm_env_sex_effect_genes)), c((length(osm_env_response_genes)-length(osm_env_response_sex_effect_genes)), (length(all_genes)-length(osm_env_response_genes)-length(non_osm_env_sex_effect_genes)))))

x
row.names(x)<- c("Sex-biased", "Unbiased")
colnames(x)<- c("Osm. env. genes", "non-osm. env. genes")
x

# The total needs to sum to length(all_genes):
sum(x)
length(all_genes)

# And now a chi square test:
chisq.test(x)

# The above shows that there are no more sex-biased osmotic environment response genes than you'd expect by chance.



```

12. Save everything:

```{r, message = FALSE}
save.image(file = "DESeq2_output_sex_effects_brain.RData")
```
