--- 
title: "Salmo smolt study"
output: plot_heatmap_and_scatterplot_brain.html
author: Joe Colgan (joscolgan)
---
  
## Introduction:  
This script generates a heatmap for significant differentially expressed genes identified across three experimental groups of brown trout that differ in their phenotype.
The script takes raw gene-level counts generated by [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html), scales counts, subsets rows by most significant genes, clusters based by expression similarity across genes and generates a heatmap.  
The output is a heatmap as .png, .jpeg and/or .pdf.

1. Load libraries: 

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2",
               "reshape2",
               "DESeq2",
               "cowplot",
               "lintr")
for (lib in libraries) {
  if (require(package = lib, character.only = TRUE)) {
    print("Successful")
  } else {
    print("Installing")
    source("https://bioconductor.org/biocLite.R")
    library(lib, character.only = TRUE)
  }
}
## Create output directory:
dir.create(path = "results")
```

2. Load data:
  
```{r, message = FALSE}
load(file = "results/DESeq2_output_liver_for_heatmap.RData")
```

3. Examine count data:
  
```{r, message = FALSE}
## Normalise counts:
normalised_counts <- t(scale(t(txi$counts),
                             center = TRUE,
                             scale = TRUE))
## Convert to dataframe:
normalised_counts_df <- as.data.frame(normalised_counts)
## Read in sample names and assign to columns of dataframes:
sample_info <- read.table(file = "data/RNAseq_Sample_File_Reduced.txt",
                          header = TRUE)

## Extract liver samples:
sample_info <- subset(x = sample_info,
                      tissue == "Liver")

## Update header information on count dataframe:
colnames(normalised_counts_df) <- sample_info$sample

## Replace 'NonSmolt' with 'Resident':
sample_info$phenotype <- gsub(pattern = "NonSmolt",
                              replacement = "Resident",
                              sample_info$phenotype)

## Update name for plotting:
## Reorder based on phenotype:
fresh_resident_samples <- sample_info[grep(pattern = "fresh",
                                      x = sample_info$environment), ]
fresh_resident_samples <- fresh_resident_samples[rev(order(fresh_resident_samples$sex)), ]

saltwater_samples <- sample_info[grep(pattern = "salt",
                                      x = sample_info$environment), ]

salt_resident_samples <- saltwater_samples[grep(pattern = "Resident",
                                      x = saltwater_samples$phenotype), ]
## Order by sex:
salt_resident_samples <- salt_resident_samples[rev(order(salt_resident_samples$sex)), ]

salt_smolt_samples <- saltwater_samples[grep(pattern = "Smolt",
                                      x = saltwater_samples$phenotype), ]
## Order by sex:
salt_smolt_samples <- salt_smolt_samples[rev(order(salt_smolt_samples$sex)), ]

#sample_info_ordered
ordered_list <- c(as.character(unlist(salt_resident_samples$sample)),
                  as.character(unlist(salt_smolt_samples$sample)),
                  as.character(unlist(fresh_resident_samples$sample)))

## Reorder the columns by phenotype:
normalised_counts_reordered <- normalised_counts_df[ , ordered_list]

## Update sample names:
resident_sw <- "R-SW"
smolt_sw    <- "S-SW"
resident_fw <- "R-FW"

## Generate new sample names for plotting:
sample_names <- c(paste(resident_sw,
                        "_",
                        "M",
                        1:nrow(subset(salt_resident_samples,
                                      sex == "Male")),
                        sep = ""),
                  paste(resident_sw, "_",
                        "F",
                        1:nrow(subset(salt_resident_samples,
                                      sex == "Female")),
                        sep = ""),
                  paste(smolt_sw,
                        "_",
                        "M",
                        1:nrow(subset(salt_smolt_samples,
                                      sex == "Male")),
                        sep = ""),
                  paste(smolt_sw,
                        "_",
                        "F",
                        1:nrow(subset(salt_smolt_samples,
                                      sex == "Female")),
                        sep = ""),  
                  paste(resident_fw,
                        "_",
                        "M",
                        1:nrow(subset(fresh_resident_samples,
                                      sex == "Male")),
                        sep = ""),
                  paste(resident_fw,
                        "_",
                        "F",
                        1:nrow(subset(fresh_resident_samples,
                                      sex == "Female")),
                        sep = ""))
                        
## Rename columns:
colnames(normalised_counts_reordered) <- sample_names
```

4. Subset genes identified as phenotype but NOT environment genes:
  
```{r, message = FALSE}
## Check length:
length(phen_not_env_genes)

## Subset rows for genes of interest:
normalised_counts_unique_degs <- subset(normalised_counts_reordered,
                                        row.names(normalised_counts_reordered) %in% phen_not_env_genes)
```

5. Extract normalised counts for genes of interest:
  
```{r, message = FALSE}
## Cluster based on euclidean distance:
dd <- hclust(dist(as.matrix(normalised_counts_unique_degs[, 1:ncol(normalised_counts_unique_degs)])))

## Reorder:
normalised_counts_ordered <- as.data.frame(normalised_counts_unique_degs[rev(dd$order), ])
normalised_counts_ordered$locus <- row.names(normalised_counts_ordered)

## Transform dataframe:
normalised_counts_ordered_melt <- melt(normalised_counts_ordered)

## Update column names:
colnames(normalised_counts_ordered_melt) <- c("gene_name",
                                              "sample",
                                              "counts")
## Update levels for plotting:
normalised_counts_ordered_melt$gene_name <-
  factor(normalised_counts_ordered_melt$gene_name,
         levels = rev(unique(c(as.character(unlist(normalised_counts_ordered_melt$gene_name))))))
```

6. Generate heatmap:
  
```{r, message = FALSE}
## Assign the name of dataframe to plot:
input <- normalised_counts_ordered_melt

## Remove underscores in sample names:
input$sample <- gsub(pattern = "_",
                     replacement = " ",
                     input$sample)

## Update levels for plotting:
input$sample <- factor(input$sample,
                       levels = unique(input$sample))

## Generate heatmap:
heatmap_plot <- ggplot(input,
                       aes(sample,
                           gene_name)) +
        geom_tile(aes(fill = counts)) +
        geom_vline(xintercept = c(0.5,
                                  8.5,
                                  15.5,
                                  24.5)) +
        scale_fill_gradient2(low = "black",
                             mid = "white",
                             high = "red") +
        ylab("") +
        xlab("") +
        theme(legend.title = element_text(size = 10,
                                          face = "bold"),
              legend.text = element_text(size = 10,
                                         face = "bold"),
              axis.text.y = element_blank(),
              axis.text.x = element_text(size = 8,
                                         face = "bold",
                                         angle = 90),
              axis.title = element_blank(),
              axis.ticks = element_blank(),
              panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              panel.background = element_blank(),
              legend.position = "top") +
        labs(fill = "Normalised gene expression")

## Print to console:
heatmap_plot
```

7. Plot heatmap:

```{r, message = FALSE}
## Save image as .png:
output  <- "results/figures_for_ms/"
ggsave(file = paste(output,
                    "heatmap_liver_phen_not_env_genes.png",
                    sep = ""),
       height = 20,
       width = 15)

## Save image as rds:
saveRDS(object = heatmap_plot,
        file = paste(output,
                    "heatmap_liver_phen_not_env_genes.rds",
                    sep = ""))

## Save image as pdf:
ggsave(file = paste(output,
                    "heatmap_liver_phen_not_env_genes.pdf",
                    sep = ""),
       height = 6,
       width = 10)
```

8. Generate scatterplot:

```{r, message = FALSE}
# Next bits are about generating a nice plot with means (as points) and SEs:
phenotype <- gsub(pattern = "_.*",
                  replacement = "",
                  sample_names)

input$phenotype<- rep(phenotype,
                      each = length(phen_not_env_genes))

sex <- gsub(pattern = "_[0-9]", replacement = "", sample_names)
sex <- gsub(pattern = ".*_",
            replacement = "",
            sex)

input$sex <- rep(sex,
                 each = length(phen_not_env_genes))

## Generate mean expression per phenotypic group and sex:
combined_df <- data.frame()
for (group in unique(input$phenotype)) {
        print(group)
        for (sex_ind in unique(input$sex)) {
                tmp_subset <- subset(input,
                                   phenotype == group &
                                   sex == sex_ind)
                tmp <- with(tmp_subset,
                            tapply(counts,
                                   gene_name,
                                   mean))
                print(sex)
                print(head(tmp))
                output_tmp <- 1:length(tmp)
                tmp_df <- as.data.frame(cbind(tmp,
                                              output_tmp))
                 tmp_df$phenotype <- group
                 tmp_df$sex <- sex_ind
                 combined_df <- rbind(combined_df,
                                      tmp_df)
        print(length(output_tmp))
        }
}

## Update levels on scatterplot:
combined_df$phenotype <- factor(combined_df$phenotype,
                                   levels = unique(combined_df$phenotype))

## Generate scatterplot:
scatterplot <- ggplot(data = combined_df,
       aes(x = tmp,
           y = output_tmp)) +
        geom_point(data = combined_df,
                   aes(fill = sex),
                   colour = "black",
                   pch = 21,
                   size = 2) +
        ylim(0, 525) +
        facet_wrap(~phenotype) +
        scale_fill_manual(values = c("green",
                          "white")) +
        theme_bw() +
        ylab("Genes") +
        xlab("Normalised gene expression") +
        theme(strip.background = element_blank(),
              #strip.text.x = element_blank(),
              axis.title = element_text(size = 10,
                                          face = "bold"),
              axis.text = element_text(size = 8,
                                       face = "bold"),
              legend.position = "top",
              legend.title = element_text(size = 10,
                                          face = "bold"),
              legend.text = element_text(size = 10,
                                         face = "bold")) +
                scale_y_continuous(position = "right",
                                   expand = c(0.01,
                            0.01)) +
        labs(fill = "Sex")

## Generate combined plot:
plot_grid(heatmap_plot + theme(legend.justification = c(0,0)),
          scatterplot + theme(legend.justification = c(0,0)),
          labels = c("A",
                     "B"),
          align = "h",
          axis = "bt",
          rel_widths = c(1.3, 1))

## Save image to file:
ggsave(file = "results/figures_for_ms/figure_2_heatmap_scatterplot.png",
       height = 8,
       width = 15)

ggsave(file = "results/figures_for_ms/figure_2_heatmap_scatterplot.pdf",
       height = 10,
       width = 15)
```

9. Run lintr:
  
```{r, message = FALSE}
lintr::lint(file = "plot_heatmap_and_scatterplot_liver.Rmd")
```
